// ngx-persian-datepicker/esm5/lib/datepicker-toggle.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/datepicker-toggle.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation, ChangeDetectorRef, } from '@angular/core';
import { MdDatepicker } from './datepicker';
import { MdDatepickerIntl } from './datepicker-intl';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
/**
 * @template D
 */
var MdDatepickerToggle = /** @class */ (function () {
    function MdDatepickerToggle(_intl, changeDetectorRef) {
        this._intl = _intl;
        this._intlChanges = _intl.changes.subscribe((/**
         * @return {?}
         */
        function () { return changeDetectorRef.markForCheck(); }));
    }
    Object.defineProperty(MdDatepickerToggle.prototype, "disabled", {
        /** Whether the toggle button is disabled. */
        get: /**
         * Whether the toggle button is disabled.
         * @return {?}
         */
        function () {
            return this._disabled === undefined ? this.datepicker.disabled : this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    MdDatepickerToggle.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._intlChanges.unsubscribe();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    MdDatepickerToggle.prototype._open = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.datepicker && !this.disabled) {
            this.datepicker.open();
            event.stopPropagation();
        }
    };
    MdDatepickerToggle.decorators = [
        { type: Component, args: [{
                    selector: 'md-datepicker-toggle, mat-datepicker-toggle',
                    template: "<button mat-icon-button type=\"button\" [attr.aria-label]=\"_intl.openCalendarLabel\"\r\n        [disabled]=\"disabled\" (click)=\"_open($event)\">\r\n  <!-- <mat-icon>\r\n    <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" width=\"100%\" height=\"100%\"\r\n         fill=\"currentColor\" style=\"vertical-align: top\" focusable=\"false\">\r\n      <path d=\"M0 0h24v24H0z\" fill=\"none\"/>\r\n      <path d=\"M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zM7 10h5v5H7z\"/>\r\n    </svg>\r\n  </mat-icon> -->\r\n</button>\r\n",
                    host: {
                        'class': 'mat-datepicker-toggle',
                    },
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    MdDatepickerToggle.ctorParameters = function () { return [
        { type: MdDatepickerIntl },
        { type: ChangeDetectorRef }
    ]; };
    MdDatepickerToggle.propDecorators = {
        datepicker: [{ type: Input, args: ['for',] }],
        disabled: [{ type: Input }]
    };
    return MdDatepickerToggle;
}());
export { MdDatepickerToggle };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdDatepickerToggle.prototype._intlChanges;
    /**
     * Datepicker instance that the button will toggle.
     * @type {?}
     */
    MdDatepickerToggle.prototype.datepicker;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerToggle.prototype._disabled;
    /** @type {?} */
    MdDatepickerToggle.prototype._intl;
}
