// ngx-persian-datepicker/esm5/lib/index.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdMonthView } from './month-view';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { A11yModule } from '@angular/cdk/a11y';
import { MdCalendarBody } from './calendar-body';
import { MdYearView } from './year-view';
import { MdDatepicker, MdDatepickerContent, MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER, } from './datepicker';
import { MdDatepickerInput } from './datepicker-input';
import { MdCalendar } from './calendar';
import { MdDatepickerToggle } from './datepicker-toggle';
import { MdDatepickerIntl } from './datepicker-intl';
import { MatButtonModule, MatButtonToggleModule, MatDialogModule, MatIconModule, MatOptionModule, MatSelectModule, DateAdapter, MAT_DATE_FORMATS } from '@angular/material';
import { MdLocaleChanger } from './locale_changer';
import { dateFormat, NativeDateAdapter } from './moment-date-adapter';
export { MdCalendar } from './calendar';
export { MdCalendarCell, MdCalendarBody } from './calendar-body';
export { MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER_FACTORY, MD_DATEPICKER_SCROLL_STRATEGY, MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER, MdDatepickerContent, MdDatepicker } from './datepicker';
export { MD_DATEPICKER_VALUE_ACCESSOR, MD_DATEPICKER_VALIDATORS, MdDatepickerInputEvent, MdDatepickerInput } from './datepicker-input';
export { MdDatepickerIntl } from './datepicker-intl';
export { MdDatepickerToggle } from './datepicker-toggle';
export { MdMonthView } from './month-view';
export { MdYearView } from './year-view';
var MatDatepickerModulePersian = /** @class */ (function () {
    function MatDatepickerModulePersian() {
    }
    MatDatepickerModulePersian.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        MatButtonModule,
                        MatButtonToggleModule,
                        MatDialogModule,
                        MatIconModule,
                        OverlayModule,
                        MatOptionModule,
                        MatSelectModule,
                        // StyleModule,
                        A11yModule,
                    ],
                    exports: [
                        MdCalendar,
                        MdCalendarBody,
                        MdDatepicker,
                        MatButtonToggleModule,
                        MdDatepickerInput,
                        MdDatepickerToggle,
                        MdMonthView,
                        MdYearView,
                    ],
                    declarations: [
                        MdCalendar,
                        MdCalendarBody,
                        MdLocaleChanger,
                        MdDatepicker,
                        MdDatepickerContent,
                        MdDatepickerInput,
                        MdDatepickerToggle,
                        MdMonthView,
                        MdYearView,
                    ],
                    providers: [
                        MdDatepickerIntl,
                        MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER,
                        { provide: DateAdapter, useClass: NativeDateAdapter }, {
                            provide: MAT_DATE_FORMATS,
                            useClass: dateFormat,
                            deps: [DateAdapter]
                        }
                    ],
                    entryComponents: [
                        MdDatepickerContent,
                    ]
                },] }
    ];
    return MatDatepickerModulePersian;
}());
export { MatDatepickerModulePersian };
