// ngx-persian-datepicker/esm5/lib/locale_changer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/locale_changer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { ChangeDetectionStrategy, Component, Inject, Optional, ViewEncapsulation } from '@angular/core';
import { createMissingDateImplError } from './datepicker-errors';
import { DateAdapter, MAT_DATE_FORMATS } from '@angular/material';
/**
 * An internal component used to display a single year in the datepicker.
 * \@docs-private
 * @template D
 */
var MdLocaleChanger = /** @class */ (function () {
    function MdLocaleChanger(_dateAdapter, _dateFormats) {
        var _this = this;
        this._dateAdapter = _dateAdapter;
        this._dateFormats = _dateFormats;
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        if (!this._dateFormats) {
            throw createMissingDateImplError('MD_DATE_FORMATS');
        }
        this.any = (/** @type {?} */ (_dateAdapter));
        this.mustDestroy = ((/** @type {?} */ (_dateFormats)))._lastChanges.subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            _this.lang = v;
        }));
    }
    /**
     * @return {?}
     */
    MdLocaleChanger.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.mustDestroy.unsubscribe();
    };
    // change locale function()
    // change locale function()
    /**
     * @return {?}
     */
    MdLocaleChanger.prototype._changeLocale = 
    // change locale function()
    /**
     * @return {?}
     */
    function () {
        this.any.changeLocale();
    };
    MdLocaleChanger.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'md-locale-changer',
                    template: "\n        <mat-button-toggle-group #group=\"matButtonToggleGroup\" (change)=\"_changeLocale()\" [value]=\"lang\">\n            <mat-button-toggle value=\"fa\">\n                \u0634\u0645\u0633\u06CC\n            </mat-button-toggle>\n            <mat-button-toggle value=\"en\">\n                \u0645\u06CC\u0644\u0627\u062F\u06CC\n            </mat-button-toggle>\n        </mat-button-toggle-group>\n    ",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".mat-button-toggle-group .mat-button-toggle{font-size:11px}.mat-button-toggle-group .mat-button-toggle .mat-button-toggle-label-content{line-height:26px}"]
                }] }
    ];
    /** @nocollapse */
    MdLocaleChanger.ctorParameters = function () { return [
        { type: DateAdapter, decorators: [{ type: Optional }] },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_FORMATS,] }] }
    ]; };
    return MdLocaleChanger;
}());
export { MdLocaleChanger };
if (false) {
    /** @type {?} */
    MdLocaleChanger.prototype.any;
    /** @type {?} */
    MdLocaleChanger.prototype.lang;
    /** @type {?} */
    MdLocaleChanger.prototype.mustDestroy;
    /** @type {?} */
    MdLocaleChanger.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdLocaleChanger.prototype._dateFormats;
}
