// ngx-persian-datepicker/esm5/lib/moment-date-adapter.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/moment-date-adapter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, LOCALE_ID, Optional } from '@angular/core';
import { DateAdapter } from '@angular/material';
import { JalaliMomentDateAdapter } from './jalali-date-adapter';
import { extendObject } from './object-extend';
import { Subject } from 'rxjs';
import { BehaviorSubject } from 'rxjs';
// TODO(mmalerba): Remove when we no longer support safari 9.
/**
 * Whether the browser supports the Intl API.
 * @type {?}
 */
var SUPPORTS_INTL_API = typeof Intl != 'undefined';
/**
 * The default month names to use if Intl API is not available.
 * @type {?}
 */
var DEFAULT_MONTH_NAMES = {
    'long': [
        'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September',
        'October', 'November', 'December'
    ],
    'short': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
    'narrow': ['J', 'F', 'M', 'A', 'M', 'J', 'J', 'A', 'S', 'O', 'N', 'D']
};
var ɵ0 = /**
 * @param {?} i
 * @return {?}
 */
function (i) { return String(i + 1); };
/**
 * The default date names to use if Intl API is not available.
 * @type {?}
 */
var DEFAULT_DATE_NAMES = range(31, (ɵ0));
/**
 * The default day of the week names to use if Intl API is not available.
 * @type {?}
 */
var DEFAULT_DAY_OF_WEEK_NAMES = {
    'long': ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
    'short': ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
    'narrow': ['S', 'M', 'T', 'W', 'T', 'F', 'S']
};
/**
 * Creates an array and fills it with values.
 * @template T
 * @param {?} length
 * @param {?} valueFunction
 * @return {?}
 */
function range(length, valueFunction) {
    /** @type {?} */
    var valuesArray = Array(length);
    for (var i = 0; i < length; i++) {
        valuesArray[i] = valueFunction(i);
    }
    return valuesArray;
}
/**
 * Adapts the native JS Date for use with cdk-based components that work with dates.
 */
var NativeDateAdapter = /** @class */ (function (_super) {
    tslib_1.__extends(NativeDateAdapter, _super);
    function NativeDateAdapter(localeId) {
        var _this = _super.call(this) || this;
        _this.lang = 'fa';
        _this.langSubject = new Subject();
        _this.JDateAdapter = new JalaliMomentDateAdapter('fa');
        /**
         * Whether to use `timeZone: 'utc'` with `Intl.DateTimeFormat` when formatting dates.
         * Without this `Intl.DateTimeFormat` sometimes chooses the wrong timeZone, which can throw off
         * the result. (e.g. in the en-US locale `new Date(1800, 7, 14).toLocaleDateString()`
         * will produce `'8/13/1800'`.
         */
        _this.useUtcForDisplay = true;
        _super.prototype.setLocale.call(_this, localeId);
        return _this;
    }
    /**
     * @return {?}
     */
    NativeDateAdapter.prototype.invalid = /**
     * @return {?}
     */
    function () {
        return new Date();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    NativeDateAdapter.prototype.toIso8601 = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return 'hi';
    };
    /**
     * @param {?} a
     * @param {?} b
     * @return {?}
     */
    NativeDateAdapter.prototype.sameDate = /**
     * @param {?} a
     * @param {?} b
     * @return {?}
     */
    function (a, b) {
        return false;
    };
    /**
     * @param {?} v
     * @return {?}
     */
    NativeDateAdapter.prototype.setLocale = /**
     * @param {?} v
     * @return {?}
     */
    function (v) {
        _super.prototype.setLocale.call(this, v);
        if (this.lang != 'fa') {
            this.lang = 'fa';
            this.langSubject.next('fa');
        }
        else {
            this.lang = 'en';
            this.langSubject.next('en');
        }
    };
    /**
     * @return {?}
     */
    NativeDateAdapter.prototype.getLang = /**
     * @return {?}
     */
    function () {
        return this.langSubject;
    };
    /**
     * @return {?}
     */
    NativeDateAdapter.prototype.changeLocale = /**
     * @return {?}
     */
    function () {
        if (this.lang != 'fa') {
            this.setLocale('fa');
        }
        else {
            this.setLocale('en');
        }
    };
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getYear = /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    function (date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getYear((/** @type {?} */ (date)));
        }
        if (date._d) {
            return date._d.getFullYear();
        }
        if (!date.getFullYear) {
            debugger;
        }
        return date.getFullYear();
    };
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getMonth = /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    function (date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getMonth((/** @type {?} */ (date)));
        }
        if (date._d) {
            return date._d.getMonth();
        }
        return date.getMonth();
    };
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getDate = /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    function (date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDate((/** @type {?} */ (date)));
        }
        if (date._d) {
            return date._d.getDate();
        }
        return date.getDate();
    };
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getDayOfWeek = /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    function (date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDayOfWeek((/** @type {?} */ (date)));
        }
        return date.getDay();
    };
    /**
     * @param {?} style
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getMonthNames = /**
     * @param {?} style
     * @param {?=} force
     * @return {?}
     */
    function (style, force) {
        var _this = this;
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getMonthNames(style);
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            var dtf_1 = new Intl.DateTimeFormat(this.locale, { month: style });
            return range(12, (/**
             * @param {?} i
             * @return {?}
             */
            function (i) { return _this._stripDirectionalityCharacters(dtf_1.format(new Date(2017, i, 1))); }));
        }
        return DEFAULT_MONTH_NAMES[style];
    };
    /**
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getDateNames = /**
     * @param {?=} force
     * @return {?}
     */
    function (force) {
        var _this = this;
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDateNames();
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            var dtf_2 = new Intl.DateTimeFormat(this.locale, { day: 'numeric' });
            return range(31, (/**
             * @param {?} i
             * @return {?}
             */
            function (i) { return _this._stripDirectionalityCharacters(dtf_2.format(new Date(2017, 0, i + 1))); }));
        }
        return DEFAULT_DATE_NAMES;
    };
    /**
     * @param {?} style
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getDayOfWeekNames = /**
     * @param {?} style
     * @param {?=} force
     * @return {?}
     */
    function (style, force) {
        var _this = this;
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDayOfWeekNames(style);
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            var dtf_3 = new Intl.DateTimeFormat(this.locale, { weekday: style });
            return range(7, (/**
             * @param {?} i
             * @return {?}
             */
            function (i) { return _this._stripDirectionalityCharacters(dtf_3.format(new Date(2017, 0, i + 1))); }));
        }
        return DEFAULT_DAY_OF_WEEK_NAMES[style];
    };
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getYearName = /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    function (date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getYearName((/** @type {?} */ (date)));
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            var dtf = new Intl.DateTimeFormat(this.locale, { year: 'numeric' });
            return this._stripDirectionalityCharacters(dtf.format(date));
        }
        return String(this.getYear(date));
    };
    /**
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getFirstDayOfWeek = /**
     * @param {?=} force
     * @return {?}
     */
    function (force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getFirstDayOfWeek();
        }
        // We can't tell using native JS Date what the first day of the week is, we default to Sunday.
        return 0;
    };
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    NativeDateAdapter.prototype.getNumDaysInMonth = /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    function (date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getNumDaysInMonth((/** @type {?} */ (date)), 'fa');
        }
        return this.getDate(this._createDateWithOverflow(this.getYear(date), this.getMonth(date) + 1, 0));
    };
    /**
     * @param {?} date
     * @return {?}
     */
    NativeDateAdapter.prototype.clone = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ ((this.JDateAdapter.clone((/** @type {?} */ (date))))));
        }
        return this.createDate(this.getYear(date), this.getMonth(date), this.getDate(date));
    };
    /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    NativeDateAdapter.prototype.createDate = /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    function (year, month, date) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ ((this.JDateAdapter.createDate(year, month, date))));
        }
        // Check for invalid month and date (except upper bound on date which we have to check after
        // creating the Date).
        if (month < 0 || month > 11) {
            throw Error("Invalid month index \"" + month + "\". Month index has to be between 0 and 11.");
        }
        if (date < 1) {
            throw Error("Invalid date \"" + date + "\". Date has to be greater than 0.");
        }
        /** @type {?} */
        var result = this._createDateWithOverflow(year, month, date);
        // Check that the date wasn't above the upper bound for the month, causing the month to overflow
        if (result.getMonth() != month) {
            throw Error("Invalid date \"" + date + "\" for month with index \"" + month + "\".");
        }
        return result;
    };
    /**
     * @return {?}
     */
    NativeDateAdapter.prototype.today = /**
     * @return {?}
     */
    function () {
        if (this.lang == 'fa') {
            return (/** @type {?} */ ((this.JDateAdapter.today())));
        }
        return new Date();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NativeDateAdapter.prototype.parse = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.parse(value, [])));
        }
        // We have no way using the native JS Date to set the parse format or locale, so we ignore these
        // parameters.
        if (typeof value == 'number') {
            return new Date(value);
        }
        return value ? new Date(Date.parse(value)) : null;
    };
    /**
     * @param {?} date
     * @param {?} displayFormat
     * @return {?}
     */
    NativeDateAdapter.prototype.format = /**
     * @param {?} date
     * @param {?} displayFormat
     * @return {?}
     */
    function (date, displayFormat) {
        if (this.lang == 'fa') {
            return this.JDateAdapter.format((/** @type {?} */ (date)), displayFormat);
        }
        if (!this.isValid(date)) {
            throw Error('NativeDateAdapter: Cannot format invalid date.');
        }
        if (SUPPORTS_INTL_API) {
            if (this.useUtcForDisplay) {
                if (date._d) {
                    date = new Date(Date.UTC(date._d.getFullYear(), date._d.getMonth(), date._d.getDate(), date._d.getHours(), date._d.getMinutes(), date._d.getSeconds(), date._d.getMilliseconds()));
                    displayFormat = extendObject({}, displayFormat, { timeZone: 'utc' });
                }
                else {
                    date = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds()));
                    displayFormat = extendObject({}, displayFormat, { timeZone: 'utc' });
                }
            }
            /** @type {?} */
            var dtf = new Intl.DateTimeFormat(this.locale, displayFormat);
            return this._stripDirectionalityCharacters(dtf.format(date));
        }
        return this._stripDirectionalityCharacters(date.toDateString());
    };
    /**
     * @param {?} date
     * @param {?} years
     * @return {?}
     */
    NativeDateAdapter.prototype.addCalendarYears = /**
     * @param {?} date
     * @param {?} years
     * @return {?}
     */
    function (date, years) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.addCalendarYears((/** @type {?} */ (date)), years)));
        }
        return this.addCalendarMonths(date, years * 12);
    };
    /**
     * @param {?} date
     * @param {?} months
     * @return {?}
     */
    NativeDateAdapter.prototype.addCalendarMonths = /**
     * @param {?} date
     * @param {?} months
     * @return {?}
     */
    function (date, months) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.addCalendarMonths((/** @type {?} */ (date)), months)));
        }
        /** @type {?} */
        var newDate = this._createDateWithOverflow(this.getYear(date), this.getMonth(date) + months, this.getDate(date));
        // It's possible to wind up in the wrong month if the original month has more days than the new
        // month. In this case we want to go to the last day of the desired month.
        // Note: the additional + 12 % 12 ensures we end up with a positive number, since JS % doesn't
        // guarantee this.
        if (this.getMonth(newDate) != ((this.getMonth(date) + months) % 12 + 12) % 12) {
            newDate = this._createDateWithOverflow(this.getYear(newDate), this.getMonth(newDate), 0);
        }
        return newDate;
    };
    /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    NativeDateAdapter.prototype.addCalendarDays = /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    function (date, days) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.addCalendarDays((/** @type {?} */ (date)), days)));
        }
        return this._createDateWithOverflow(this.getYear(date), this.getMonth(date), this.getDate(date) + days);
    };
    /**
     * @param {?} date
     * @return {?}
     */
    NativeDateAdapter.prototype.getISODateString = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        if (this.lang == 'fa') {
            return this.JDateAdapter.getISODateString((/** @type {?} */ (date)));
        }
        return [
            date.getUTCFullYear(),
            this._2digit(date.getUTCMonth() + 1),
            this._2digit(date.getUTCDate())
        ].join('-');
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    NativeDateAdapter.prototype.isDateInstance = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        return obj instanceof Date;
    };
    /**
     * @param {?} date
     * @return {?}
     */
    NativeDateAdapter.prototype.isValid = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        if (date._d) {
            return !isNaN(date._d.getTime());
        }
        return !isNaN(date.getTime());
    };
    /** Creates a date but allows the month and date to overflow. */
    /**
     * Creates a date but allows the month and date to overflow.
     * @private
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    NativeDateAdapter.prototype._createDateWithOverflow = /**
     * Creates a date but allows the month and date to overflow.
     * @private
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    function (year, month, date) {
        /** @type {?} */
        var result = new Date(year, month, date);
        // We need to correct for the fact that JS native Date treats years in range [0, 99] as
        // abbreviations for 19xx.
        if (year >= 0 && year < 100) {
            result.setFullYear(this.getYear(result) - 1900);
        }
        return result;
    };
    /**
     * Pads a number to make it two digits.
     * @param n The number to pad.
     * @returns The padded number.
     */
    /**
     * Pads a number to make it two digits.
     * @private
     * @param {?} n The number to pad.
     * @return {?} The padded number.
     */
    NativeDateAdapter.prototype._2digit = /**
     * Pads a number to make it two digits.
     * @private
     * @param {?} n The number to pad.
     * @return {?} The padded number.
     */
    function (n) {
        return ('00' + n).slice(-2);
    };
    /**
     * Strip out unicode LTR and RTL characters. Edge and IE insert these into formatted dates while
     * other browsers do not. We remove them to make output consistent and because they interfere with
     * date parsing.
     * @param str The string to strip direction characters from.
     * @returns The stripped string.
     */
    /**
     * Strip out unicode LTR and RTL characters. Edge and IE insert these into formatted dates while
     * other browsers do not. We remove them to make output consistent and because they interfere with
     * date parsing.
     * @private
     * @param {?} str The string to strip direction characters from.
     * @return {?} The stripped string.
     */
    NativeDateAdapter.prototype._stripDirectionalityCharacters = /**
     * Strip out unicode LTR and RTL characters. Edge and IE insert these into formatted dates while
     * other browsers do not. We remove them to make output consistent and because they interfere with
     * date parsing.
     * @private
     * @param {?} str The string to strip direction characters from.
     * @return {?} The stripped string.
     */
    function (str) {
        return str.replace(/[\u200e\u200f]/g, '');
    };
    NativeDateAdapter.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    NativeDateAdapter.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    return NativeDateAdapter;
}(DateAdapter));
export { NativeDateAdapter };
if (false) {
    /** @type {?} */
    NativeDateAdapter.prototype.lang;
    /** @type {?} */
    NativeDateAdapter.prototype.langSubject;
    /** @type {?} */
    NativeDateAdapter.prototype.JDateAdapter;
    /**
     * Whether to use `timeZone: 'utc'` with `Intl.DateTimeFormat` when formatting dates.
     * Without this `Intl.DateTimeFormat` sometimes chooses the wrong timeZone, which can throw off
     * the result. (e.g. in the en-US locale `new Date(1800, 7, 14).toLocaleDateString()`
     * will produce `'8/13/1800'`.
     * @type {?}
     */
    NativeDateAdapter.prototype.useUtcForDisplay;
}
var dateFormat = /** @class */ (function () {
    function dateFormat(_dateAdapter) {
        var _this = this;
        this.parse = {
            dateInput: 'l',
        };
        this.display = {
            dateInput: 'jYYYY/jMM/jDD',
            monthYearLabel: 'jYYYY jMMMM',
            dateA11yLabel: 'jYYYY/jMM/jDD',
            monthYearA11yLabel: 'jYYYY jMMMM'
        };
        this._getChanges = new Subject();
        this._lastChanges = new BehaviorSubject('fa');
        /** @type {?} */
        var dateAdapter = (/** @type {?} */ (_dateAdapter));
        dateAdapter.getLang().subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            if (v == 'fa') {
                _this.parse = {
                    dateInput: 'l',
                };
                _this.display = {
                    dateInput: 'jYYYY/jMM/jDD',
                    monthYearLabel: 'jYYYY jMMMM',
                    dateA11yLabel: 'jYYYY/jMM/jDD',
                    monthYearA11yLabel: 'jYYYY jMMMM'
                };
            }
            else {
                _this.parse = {
                    dateInput: null,
                };
                _this.display = {
                    dateInput: { year: 'numeric', month: 'numeric', day: 'numeric' },
                    monthYearLabel: { year: 'numeric', month: 'short' },
                    dateA11yLabel: { year: 'numeric', month: 'long', day: 'numeric' },
                    monthYearA11yLabel: { year: 'numeric', month: 'long' },
                };
            }
            _this._lastChanges.next(v);
            _this._getChanges.next(v);
        }), (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
        }), (/**
         * @return {?}
         */
        function () {
        }));
    }
    /**
     * @return {?}
     */
    dateFormat.prototype.getChanges = /**
     * @return {?}
     */
    function () {
        return this._getChanges;
    };
    return dateFormat;
}());
export { dateFormat };
if (false) {
    /** @type {?} */
    dateFormat.prototype.parse;
    /** @type {?} */
    dateFormat.prototype.display;
    /** @type {?} */
    dateFormat.prototype._getChanges;
    /** @type {?} */
    dateFormat.prototype._lastChanges;
}
export { ɵ0 };
