// ngx-persian-datepicker/fesm2015/ngx-persian-datepicker.js
import { EventEmitter, Component, ViewEncapsulation, ChangeDetectionStrategy, Input, Output, Optional, Inject, ChangeDetectorRef, Injectable, ElementRef, NgZone, InjectionToken, ViewChild, ViewContainerRef, forwardRef, Directive, Renderer2, LOCALE_ID, NgModule } from '@angular/core';
import { DateAdapter, MAT_DATE_FORMATS, MatDialog, MatFormField, MatButtonModule, MatButtonToggleModule, MatDialogModule, MatIconModule, MatOptionModule, MatSelectModule } from '@angular/material';
import { DOCUMENT, CommonModule } from '@angular/common';
import { Overlay, OverlayConfig, OverlayModule } from '@angular/cdk/overlay';
import { A11yModule } from '@angular/cdk/a11y';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ComponentPortal } from '@angular/cdk/portal';
import { Directionality } from '@angular/cdk/bidi';
import { ENTER, PAGE_DOWN, PAGE_UP, END, HOME, DOWN_ARROW, UP_ARROW, RIGHT_ARROW, LEFT_ARROW, ESCAPE } from '@angular/cdk/keycodes';
import { Subject, BehaviorSubject } from 'rxjs';
import { merge, first } from 'rxjs/operators';
import { NG_VALUE_ACCESSOR, NG_VALIDATORS, Validators } from '@angular/forms';
import * as jMoment from 'moment-jalaali';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/calendar-body.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * An internal class that represents the data corresponding to a single calendar cell.
 * \@docs-private
 */
class MdCalendarCell {
    /**
     * @param {?} value
     * @param {?} displayValue
     * @param {?} ariaLabel
     * @param {?} enabled
     * @param {?} other
     */
    constructor(value, displayValue, ariaLabel, enabled, other) {
        this.value = value;
        this.displayValue = displayValue;
        this.ariaLabel = ariaLabel;
        this.enabled = enabled;
        this.other = other;
    }
}
if (false) {
    /** @type {?} */
    MdCalendarCell.prototype.value;
    /** @type {?} */
    MdCalendarCell.prototype.displayValue;
    /** @type {?} */
    MdCalendarCell.prototype.ariaLabel;
    /** @type {?} */
    MdCalendarCell.prototype.enabled;
    /** @type {?} */
    MdCalendarCell.prototype.other;
}
/**
 * An internal component used to display calendar data in a table.
 * \@docs-private
 */
class MdCalendarBody {
    constructor() {
        /**
         * The number of columns in the table.
         */
        this.numCols = 7;
        /**
         * Whether to allow selection of disabled cells.
         */
        this.allowDisabledSelection = false;
        /**
         * The cell number of the active cell in the table.
         */
        this.activeCell = 0;
        /**
         * The aspect ratio (width / height) to use for the cells in the table. This aspect ratio will be
         * maintained even as the table resizes.
         */
        this.cellAspectRatio = 1;
        /**
         * Emits when a new value is selected.
         */
        this.selectedValueChange = new EventEmitter();
    }
    /**
     * @param {?} cell
     * @return {?}
     */
    _cellClicked(cell) {
        if (!this.allowDisabledSelection && !cell.enabled) {
            return;
        }
        this.selectedValueChange.emit(cell.value);
    }
    /**
     * The number of blank cells to put at the beginning for the first row.
     * @return {?}
     */
    get _firstRowOffset() {
        return this.rows && this.rows.length && this.rows[0].length ? this.numCols - this.rows[0].length : 0;
    }
    /**
     * @param {?} rowIndex
     * @param {?} colIndex
     * @return {?}
     */
    _isActiveCell(rowIndex, colIndex) {
        /** @type {?} */
        let cellNumber = rowIndex * this.numCols + colIndex;
        // Account for the fact that the first row may not have as many cells.
        if (rowIndex) {
            cellNumber -= this._firstRowOffset;
        }
        return cellNumber == this.activeCell;
    }
}
MdCalendarBody.decorators = [
    { type: Component, args: [{
                moduleId: module.id,
                selector: '[md-calendar-body], [mat-calendar-body]',
                template: "<!--\r\n  If there's not enough space in the first row, create a separate label row. We mark this row as\r\n  aria-hidden because we don't want it to be read out as one of the weeks in the month.\r\n-->\r\n<!--<tr *ngIf=\"_firstRowOffset < labelMinRequiredCells\" aria-hidden=\"true\">-->\r\n  <!--<td class=\"mat-calendar-body-label\"-->\r\n      <!--[attr.colspan]=\"numCols\"-->\r\n      <!--[style.paddingTop.%]=\"50 * cellAspectRatio / numCols\"-->\r\n      <!--[style.paddingBottom.%]=\"50 * cellAspectRatio / numCols\">-->\r\n    <!--{{label}}-->\r\n  <!--</td>-->\r\n<!--</tr>-->\r\n\r\n<!-- Create the first row separately so we can include a special spacer cell. -->\r\n<tr *ngFor=\"let row of rows; let rowIndex = index\" role=\"row\">\r\n  <!--\r\n    We mark this cell as aria-hidden so it doesn't get read out as one of the days in the week.\r\n    The aspect ratio of the table cells is maintained by setting the top and bottom padding as a\r\n    percentage of the width (a variant of the trick described here:\r\n    https://www.w3schools.com/howto/howto_css_aspect_ratio.asp).\r\n  -->\r\n  <td *ngIf=\"rowIndex === 0 && _firstRowOffset\"\r\n      aria-hidden=\"true\"\r\n      class=\"mat-calendar-body-label\"\r\n      [attr.colspan]=\"_firstRowOffset\"\r\n      [style.paddingTop.%]=\"50 * cellAspectRatio / numCols\"\r\n      [style.paddingBottom.%]=\"50 * cellAspectRatio / numCols\">\r\n    <!--{{_firstRowOffset >= labelMinRequiredCells ? label : ''}}-->\r\n  </td>\r\n  <td *ngFor=\"let item of row; let colIndex = index\"\r\n      role=\"gridcell\"\r\n      class=\"mat-calendar-body-cell\"\r\n      [tabindex]=\"_isActiveCell(rowIndex, colIndex) ? 0 : -1\"\r\n      [class.mat-calendar-body-disabled]=\"!item.enabled\"\r\n      [class.mat-calendar-body-active]=\"_isActiveCell(rowIndex, colIndex)\"\r\n      [attr.aria-label]=\"item.ariaLabel\"\r\n      [attr.aria-disabled]=\"!item.enabled || null\"\r\n      (click)=\"_cellClicked(item)\"\r\n      [style.width.%]=\"100 / numCols\"\r\n      [style.paddingTop.%]=\"50 * cellAspectRatio / numCols\"\r\n      [style.paddingBottom.%]=\"50 * cellAspectRatio / numCols\">\r\n    <div class=\"mat-calendar-body-cell-content\"\r\n         [class.mat-calendar-body-selected]=\"selectedValue === item.value\"\r\n         [class.subed]=\"subedValue === item.value\"\r\n         [class.mat-calendar-body-today]=\"todayValue === item.value\">\r\n      {{item.displayValue}}\r\n        <div class=\"superSmall mdc-theme--primary-light\">\r\n            {{item?.other}}\r\n        </div>\r\n    </div>\r\n  </td>\r\n</tr>\r\n",
                host: {
                    'class': 'mat-calendar-body',
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".mat-calendar-body{min-width:224px;font-size:14px}.superSmall{position:absolute;bottom:3px;left:0;width:100%;text-align:center;color:#d6ac97;font-size:12px}.wide-calendar{width:592px;height:auto}.wide-calendar .mat-calendar-content{display:flex}.wide-calendar .mat-calendar-content md-year-view{max-width:300px;margin:0 auto;width:100%}.wide-calendar .mat-calendar-content md-month-view.ng-star-inserted:first-child:after{content:\"\";background:#bbb;position:absolute;left:-10px;height:calc(97% - 20px);top:20px;width:1px}.wide-calendar .mat-calendar-content md-month-view{position:relative;width:50%;margin:0 10px}.wide-calendar .mat-calendar-content md-month-view .mat-calendar-table thead th:after{left:0;right:0}.mat-calendar-body-label{height:0;line-height:0;text-align:left;padding-left:4.7142857143%;padding-right:4.7142857143%}.mat-calendar-body-cell{position:relative;height:0;line-height:0;text-align:center;outline:0;cursor:pointer}.mat-calendar-body-disabled{cursor:default}.mat-calendar-body-cell-content{padding-bottom:5px;position:absolute;top:5%;left:5%;display:flex;align-items:center;justify-content:center;box-sizing:border-box;width:90%;height:90%;line-height:11px;border-width:1px;border-style:solid;border-radius:999px}[dir=rtl] .mat-calendar-body-label{text-align:right}"]
            }] }
];
MdCalendarBody.propDecorators = {
    label: [{ type: Input }],
    rows: [{ type: Input }],
    todayValue: [{ type: Input }],
    selectedValue: [{ type: Input }],
    subedValue: [{ type: Input }],
    labelMinRequiredCells: [{ type: Input }],
    numCols: [{ type: Input }],
    allowDisabledSelection: [{ type: Input }],
    activeCell: [{ type: Input }],
    cellAspectRatio: [{ type: Input }],
    selectedValueChange: [{ type: Output }]
};
if (false) {
    /**
     * The label for the table. (e.g. "Jan 2017").
     * @type {?}
     */
    MdCalendarBody.prototype.label;
    /**
     * The cells to display in the table.
     * @type {?}
     */
    MdCalendarBody.prototype.rows;
    /**
     * The value in the table that corresponds to today.
     * @type {?}
     */
    MdCalendarBody.prototype.todayValue;
    /**
     * The value in the table that is currently selected.
     * @type {?}
     */
    MdCalendarBody.prototype.selectedValue;
    /** @type {?} */
    MdCalendarBody.prototype.subedValue;
    /**
     * The minimum number of free cells needed to fit the label in the first row.
     * @type {?}
     */
    MdCalendarBody.prototype.labelMinRequiredCells;
    /**
     * The number of columns in the table.
     * @type {?}
     */
    MdCalendarBody.prototype.numCols;
    /**
     * Whether to allow selection of disabled cells.
     * @type {?}
     */
    MdCalendarBody.prototype.allowDisabledSelection;
    /**
     * The cell number of the active cell in the table.
     * @type {?}
     */
    MdCalendarBody.prototype.activeCell;
    /**
     * The aspect ratio (width / height) to use for the cells in the table. This aspect ratio will be
     * maintained even as the table resizes.
     * @type {?}
     */
    MdCalendarBody.prototype.cellAspectRatio;
    /**
     * Emits when a new value is selected.
     * @type {?}
     */
    MdCalendarBody.prototype.selectedValueChange;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/datepicker-errors.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * \@docs-private
 * @param {?} provider
 * @return {?}
 */
function createMissingDateImplError(provider) {
    return Error(`MdDatepicker: No provider found for ${provider}. You must import one of the following ` +
        `modules at your application root: MdNativeDateModule, or provide a custom implementation.`);
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/month-view.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DAYS_PER_WEEK = 7;
/**
 * An internal component used to display a single month in the datepicker.
 * \@docs-private
 * @template D
 */
class MdMonthView {
    /**
     * @param {?} _dateAdapter
     * @param {?} _dateFormats
     * @param {?} ref
     */
    constructor(_dateAdapter, _dateFormats, ref) {
        this._dateAdapter = _dateAdapter;
        this._dateFormats = _dateFormats;
        /**
         * Emits when a new date is selected.
         */
        this.selectedChange = new EventEmitter();
        /**
         * Emits when any date is selected.
         */
        this.userSelection = new EventEmitter();
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        if (!this._dateFormats) {
            throw createMissingDateImplError('MD_DATE_FORMATS');
        }
        this._activeDate = this._dateAdapter.today();
        /** @type {?} */
        let any = (/** @type {?} */ (_dateFormats));
        this.mustDestroy = any.getChanges().subscribe((/**
         * @param {?} v
         * @return {?}
         */
        (v) => {
            this._init();
            ref.detectChanges();
        }), (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
        }), (/**
         * @return {?}
         */
        () => {
        }));
        this.mustDestroy2 = any._lastChanges.subscribe((/**
         * @param {?} v
         * @return {?}
         */
        (v) => {
            this.lang = v;
        }));
    }
    /**
     * The date to display in this month view (everything other than the month and year is ignored).
     * @return {?}
     */
    get activeDate() {
        return this._activeDate;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set activeDate(value) {
        /** @type {?} */
        let oldActiveDate = this._activeDate;
        this._activeDate = value || this._dateAdapter.today();
        if (!this._hasSameMonthAndYear(oldActiveDate, this._activeDate)) {
            this._init();
        }
    }
    /**
     * The currently selected date.
     * @return {?}
     */
    get selected() {
        return this._selected;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set selected(value) {
        this._selected = value;
        this._selectedDate = this._getDateInCurrentMonth(this.selected);
    }
    /**
     * @return {?}
     */
    get subedDate() {
        return this._subed;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set subedDate(value) {
        this._subed = value;
        this._subedDate = this._getDateInCurrentMonth(this.subedDate);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.mustDestroy.unsubscribe();
        this.mustDestroy2.unsubscribe();
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._init();
    }
    /**
     * Handles when a new date is selected.
     * @param {?} date
     * @return {?}
     */
    _dateSelected(date) {
        /** @type {?} */
        const selectedYear = this._dateAdapter.getYear(this.activeDate);
        /** @type {?} */
        const selectedMonth = this._dateAdapter.getMonth(this.activeDate);
        /** @type {?} */
        const selectedDate = this._dateAdapter.createDate(selectedYear, selectedMonth, date);
        this.selectedChange.emit(selectedDate);
        this.userSelection.emit();
    }
    /**
     * Initializes this month view.
     * @private
     * @return {?}
     */
    _init() {
        /** @type {?} */
        const firstDayOfWeek = this._dateAdapter.getFirstDayOfWeek();
        /** @type {?} */
        const narrowWeekdays = this._dateAdapter.getDayOfWeekNames('narrow');
        /** @type {?} */
        const longWeekdays = this._dateAdapter.getDayOfWeekNames('long');
        // Rotate the labels for days of the week based on the configured first day of the week.
        /** @type {?} */
        let weekdays = longWeekdays.map((/**
         * @param {?} long
         * @param {?} i
         * @return {?}
         */
        (long, i) => {
            return { long, narrow: narrowWeekdays[i] };
        }));
        this._weekdays = weekdays.slice(firstDayOfWeek).concat(weekdays.slice(0, firstDayOfWeek));
        this._selectedDate = this._getDateInCurrentMonth(this.selected);
        this._todayDate = this._getDateInCurrentMonth(this._dateAdapter.today());
        this._yearNumber = this._dateAdapter.getYear(this.activeDate);
        this._monthLabel =
            this._dateAdapter.getMonthNames('short')[this._dateAdapter.getMonth(this.activeDate)]
                .toLocaleUpperCase();
        /** @type {?} */
        let firstOfMonth = this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), this._dateAdapter.getMonth(this.activeDate), 1);
        this._firstWeekOffset =
            (DAYS_PER_WEEK + this._dateAdapter.getDayOfWeek(firstOfMonth) -
                this._dateAdapter.getFirstDayOfWeek()) % DAYS_PER_WEEK;
        this._createWeekCells();
    }
    /**
     * Creates MdCalendarCells for the dates in this month.
     * @private
     * @return {?}
     */
    _createWeekCells() {
        /** @type {?} */
        let daysInMonthFa;
        /** @type {?} */
        let dateNamesFa;
        /** @type {?} */
        let daysInMonth = this._dateAdapter.getNumDaysInMonth(this._activeDate);
        /** @type {?} */
        let dateNames = this._dateAdapter.getDateNames();
        if (((/** @type {?} */ (this._dateAdapter))).JDateAdapter) {
            /** @type {?} */
            let JAdapter = ((/** @type {?} */ (this._dateAdapter))).JDateAdapter;
            daysInMonthFa = JAdapter.getNumDaysInMonth(this._activeDate);
            dateNamesFa = JAdapter.getDateNames();
        }
        this._weeks = [[]];
        for (let i = 0, cell = this._firstWeekOffset; i < daysInMonth; i++, cell++) {
            /** @type {?} */
            let ariaLabelFa;
            /** @type {?} */
            let dateFa;
            /** @type {?} */
            let dateEn;
            if (cell == DAYS_PER_WEEK) {
                this._weeks.push([]);
                cell = 0;
            }
            /** @type {?} */
            let date = this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), this._dateAdapter.getMonth(this.activeDate), i + 1);
            /** @type {?} */
            let enabled = !this.dateFilter ||
                this.dateFilter(date);
            if (((/** @type {?} */ (this._dateAdapter))).JDateAdapter) {
                /** @type {?} */
                let JAdapter = ((/** @type {?} */ (this._dateAdapter))).JDateAdapter;
                dateFa = JAdapter.getDate(date);
                dateEn = ((/** @type {?} */ (this._dateAdapter.getDate)))(date, true);
                // ariaLabelFa = JAdapter.format(dateFa, '');
            }
            /** @type {?} */
            let ariaLabel = this._dateAdapter.format(date, this._dateFormats.display.dateA11yLabel);
            if (this.lang == 'en') {
                this._weeks[this._weeks.length - 1]
                    .push(new MdCalendarCell(i + 1, dateNames[i], ariaLabel, enabled, dateFa));
            }
            else {
                this._weeks[this._weeks.length - 1]
                    .push(new MdCalendarCell(i + 1, dateNames[i], ariaLabel, enabled, dateEn));
            }
        }
    }
    /**
     * Gets the date in this month that the given Date falls on.
     * Returns null if the given Date is in another month.
     * @private
     * @param {?} date
     * @return {?}
     */
    _getDateInCurrentMonth(date) {
        return this._hasSameMonthAndYear(date, this.activeDate) ?
            this._dateAdapter.getDate(date) : null;
    }
    /**
     * Checks whether the 2 dates are non-null and fall within the same month of the same year.
     * @private
     * @param {?} d1
     * @param {?} d2
     * @return {?}
     */
    _hasSameMonthAndYear(d1, d2) {
        return !!(d1 && d2 && this._dateAdapter.getMonth(d1) == this._dateAdapter.getMonth(d2) &&
            this._dateAdapter.getYear(d1) == this._dateAdapter.getYear(d2));
    }
}
MdMonthView.decorators = [
    { type: Component, args: [{
                moduleId: module.id,
                selector: 'md-month-view',
                template: "<div class=\"mlabel\">{{_monthLabel}}<span>{{_yearNumber}}</span></div>\r\n<table class=\"mat-calendar-table\">\r\n  <thead class=\"mat-calendar-table-header\">\r\n    <tr><th *ngFor=\"let day of this?._weekdays\" [attr.aria-label]=\"day.long\">{{day.narrow}}</th></tr>\r\n    <tr><th class=\"mat-calendar-table-header-divider\" colspan=\"7\" aria-hidden=\"true\"></th></tr>\r\n  </thead>\r\n  <tbody md-calendar-body\r\n         role=\"grid\"\r\n         [label]=\"_monthLabel\"\r\n         [rows]=\"_weeks\"\r\n         [todayValue]=\"_todayDate\"\r\n         [selectedValue]=\"_selectedDate\"\r\n         [labelMinRequiredCells]=\"3\"\r\n         [activeCell]=\"_dateAdapter.getDate(activeDate) - 1\"\r\n         [subedValue]=\"_subedDate\"\r\n         (selectedValueChange)=\"_dateSelected($event)\">\r\n  </tbody>\r\n</table>\r\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
MdMonthView.ctorParameters = () => [
    { type: DateAdapter, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_FORMATS,] }] },
    { type: ChangeDetectorRef }
];
MdMonthView.propDecorators = {
    activeDate: [{ type: Input }],
    selected: [{ type: Input }],
    subedDate: [{ type: Input }],
    dateFilter: [{ type: Input }],
    selectedChange: [{ type: Output }],
    userSelection: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdMonthView.prototype._activeDate;
    /**
     * @type {?}
     * @private
     */
    MdMonthView.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    MdMonthView.prototype._subed;
    /**
     * A function used to filter which dates are selectable.
     * @type {?}
     */
    MdMonthView.prototype.dateFilter;
    /**
     * Emits when a new date is selected.
     * @type {?}
     */
    MdMonthView.prototype.selectedChange;
    /**
     * Emits when any date is selected.
     * @type {?}
     */
    MdMonthView.prototype.userSelection;
    /**
     * The label for this month (e.g. "January 2017").
     * @type {?}
     */
    MdMonthView.prototype._monthLabel;
    /** @type {?} */
    MdMonthView.prototype._yearNumber;
    /**
     * Grid of calendar cells representing the dates of the month.
     * @type {?}
     */
    MdMonthView.prototype._weeks;
    /**
     * The number of blank cells in the first row before the 1st of the month.
     * @type {?}
     */
    MdMonthView.prototype._firstWeekOffset;
    /**
     * The date of the month that the currently selected Date falls on.
     * Null if the currently selected Date is in another month.
     * @type {?}
     */
    MdMonthView.prototype._selectedDate;
    /** @type {?} */
    MdMonthView.prototype._subedDate;
    /** @type {?} */
    MdMonthView.prototype.mustDestroy;
    /** @type {?} */
    MdMonthView.prototype.mustDestroy2;
    /**
     * The date of the month that today falls on. Null if today is in another month.
     * @type {?}
     */
    MdMonthView.prototype._todayDate;
    /**
     * The names of the weekdays.
     * @type {?}
     */
    MdMonthView.prototype._weekdays;
    /** @type {?} */
    MdMonthView.prototype.lang;
    /** @type {?} */
    MdMonthView.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdMonthView.prototype._dateFormats;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/year-view.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * An internal component used to display a single year in the datepicker.
 * \@docs-private
 * @template D
 */
class MdYearView {
    /**
     * @param {?} _dateAdapter
     * @param {?} _dateFormats
     * @param {?} ref
     */
    constructor(_dateAdapter, _dateFormats, ref) {
        this._dateAdapter = _dateAdapter;
        this._dateFormats = _dateFormats;
        /**
         * Emits when a new month is selected.
         */
        this.selectedChange = new EventEmitter();
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        if (!this._dateFormats) {
            throw createMissingDateImplError('MD_DATE_FORMATS');
        }
        this.mustDestroy = this._dateFormats._getChanges.subscribe((/**
         * @param {?} v
         * @return {?}
         */
        (v) => {
            this._init();
            ref.detectChanges();
        }));
        this._activeDate = this._dateAdapter.today();
    }
    /**
     * The date to display in this year view (everything other than the year is ignored).
     * @return {?}
     */
    get activeDate() {
        return this._activeDate;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set activeDate(value) {
        /** @type {?} */
        let oldActiveDate = this._activeDate;
        this._activeDate = value || this._dateAdapter.today();
        if (this._dateAdapter.getYear(oldActiveDate) != this._dateAdapter.getYear(this._activeDate)) {
            this._init();
        }
    }
    /**
     * The currently selected date.
     * @return {?}
     */
    get selected() {
        return this._selected;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set selected(value) {
        this._selected = value;
        this._selectedMonth = this._getMonthInCurrentYear(this.selected);
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._init();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.mustDestroy.unsubscribe();
    }
    /**
     * Handles when a new month is selected.
     * @param {?} month
     * @return {?}
     */
    _monthSelected(month) {
        /** @type {?} */
        let daysInMonth = this._dateAdapter.getNumDaysInMonth(this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), month, 1));
        this.selectedChange.emit(this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), month, Math.min(this._dateAdapter.getDate(this.activeDate), daysInMonth)));
    }
    /**
     * Initializes this month view.
     * @private
     * @return {?}
     */
    _init() {
        this._selectedMonth = this._getMonthInCurrentYear(this.selected);
        this._todayMonth = this._getMonthInCurrentYear(this._dateAdapter.today());
        this._yearLabel = this._dateAdapter.getYearName(this.activeDate);
        /** @type {?} */
        let monthNames = this._dateAdapter.getMonthNames('short');
        // First row of months only contains 5 elements so we can fit the year label on the same row.
        this._months = [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]].map((/**
         * @param {?} row
         * @return {?}
         */
        row => row.map((/**
         * @param {?} month
         * @return {?}
         */
        month => this._createCellForMonth(month, monthNames[month])))));
    }
    /**
     * Gets the month in this year that the given Date falls on.
     * Returns null if the given Date is in another year.
     * @private
     * @param {?} date
     * @return {?}
     */
    _getMonthInCurrentYear(date) {
        return date && this._dateAdapter.getYear(date) == this._dateAdapter.getYear(this.activeDate) ?
            this._dateAdapter.getMonth(date) : null;
    }
    /**
     * Creates an MdCalendarCell for the given month.
     * @private
     * @param {?} month
     * @param {?} monthName
     * @return {?}
     */
    _createCellForMonth(month, monthName) {
        /** @type {?} */
        let ariaLabel = this._dateAdapter.format(this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), month, 1), this._dateFormats.display.monthYearA11yLabel);
        return new MdCalendarCell(month, monthName.toLocaleUpperCase(), ariaLabel, this._isMonthEnabled(month), null);
    }
    /**
     * Whether the given month is enabled.
     * @private
     * @param {?} month
     * @return {?}
     */
    _isMonthEnabled(month) {
        if (!this.dateFilter) {
            return true;
        }
        /** @type {?} */
        let firstOfMonth = this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), month, 1);
        // If any date in the month is enabled count the month as enabled.
        for (let date = firstOfMonth; this._dateAdapter.getMonth(date) == month; date = this._dateAdapter.addCalendarDays(date, 1)) {
            if (this.dateFilter(date)) {
                return true;
            }
        }
        return false;
    }
}
MdYearView.decorators = [
    { type: Component, args: [{
                moduleId: module.id,
                selector: 'md-year-view, mat-year-view',
                template: "<table class=\"mat-calendar-table\">\r\n  <thead class=\"mat-calendar-table-header\">\r\n    <tr><th class=\"mat-calendar-table-header-divider\" colspan=\"4\"></th></tr>\r\n  </thead>\r\n  <tbody md-calendar-body\r\n         role=\"grid\"\r\n         allowDisabledSelection=\"true\"\r\n         [label]=\"_yearLabel\"\r\n         [rows]=\"_months\"\r\n         [todayValue]=\"_todayMonth\"\r\n         [selectedValue]=\"_selectedMonth\"\r\n         [labelMinRequiredCells]=\"2\"\r\n         [numCols]=\"4\"\r\n         [cellAspectRatio]=\"4 / 7\"\r\n         [activeCell]=\"_dateAdapter.getMonth(activeDate)\"\r\n         (selectedValueChange)=\"_monthSelected($event)\">\r\n  </tbody>\r\n</table>\r\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
MdYearView.ctorParameters = () => [
    { type: DateAdapter, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_FORMATS,] }] },
    { type: ChangeDetectorRef }
];
MdYearView.propDecorators = {
    activeDate: [{ type: Input }],
    selected: [{ type: Input }],
    dateFilter: [{ type: Input }],
    selectedChange: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdYearView.prototype._activeDate;
    /**
     * @type {?}
     * @private
     */
    MdYearView.prototype._selected;
    /**
     * A function used to filter which dates are selectable.
     * @type {?}
     */
    MdYearView.prototype.dateFilter;
    /**
     * Emits when a new month is selected.
     * @type {?}
     */
    MdYearView.prototype.selectedChange;
    /**
     * Grid of calendar cells representing the months of the year.
     * @type {?}
     */
    MdYearView.prototype._months;
    /**
     * The label for this year (e.g. "2017").
     * @type {?}
     */
    MdYearView.prototype._yearLabel;
    /**
     * The month in this year that today falls on. Null if today is in a different year.
     * @type {?}
     */
    MdYearView.prototype._todayMonth;
    /**
     * The month in this year that the selected Date falls on.
     * Null if the selected Date is in a different year.
     * @type {?}
     */
    MdYearView.prototype._selectedMonth;
    /** @type {?} */
    MdYearView.prototype.mustDestroy;
    /** @type {?} */
    MdYearView.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdYearView.prototype._dateFormats;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/datepicker-intl.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Datepicker data that requires internationalization.
 */
class MdDatepickerIntl {
    constructor() {
        /**
         * Stream that emits whenever the labels here are changed. Use this to notify
         * components if the labels have changed after initialization.
         */
        this.changes = new Subject();
        /**
         * A label for the calendar popup (used by screen readers).
         */
        this.calendarLabel = 'Calendar';
        /**
         * A label for the button used to open the calendar popup (used by screen readers).
         */
        this.openCalendarLabel = 'Open calendar';
        /**
         * A label for the previous month button (used by screen readers).
         */
        this.prevMonthLabel = 'Previous month';
        /**
         * A label for the next month button (used by screen readers).
         */
        this.nextMonthLabel = 'Next month';
        /**
         * A label for the previous year button (used by screen readers).
         */
        this.prevYearLabel = 'Previous year';
        /**
         * A label for the next year button (used by screen readers).
         */
        this.nextYearLabel = 'Next year';
        /**
         * A label for the 'switch to month view' button (used by screen readers).
         */
        this.switchToMonthViewLabel = 'Change to month view';
        /**
         * A label for the 'switch to year view' button (used by screen readers).
         */
        this.switchToYearViewLabel = 'Change to year view';
    }
}
MdDatepickerIntl.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * Stream that emits whenever the labels here are changed. Use this to notify
     * components if the labels have changed after initialization.
     * @type {?}
     */
    MdDatepickerIntl.prototype.changes;
    /**
     * A label for the calendar popup (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.calendarLabel;
    /**
     * A label for the button used to open the calendar popup (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.openCalendarLabel;
    /**
     * A label for the previous month button (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.prevMonthLabel;
    /**
     * A label for the next month button (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.nextMonthLabel;
    /**
     * A label for the previous year button (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.prevYearLabel;
    /**
     * A label for the next year button (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.nextYearLabel;
    /**
     * A label for the 'switch to month view' button (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.switchToMonthViewLabel;
    /**
     * A label for the 'switch to year view' button (used by screen readers).
     * @type {?}
     */
    MdDatepickerIntl.prototype.switchToYearViewLabel;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/calendar.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * A calendar that is used as part of the datepicker.
 * \@docs-private
 * @template D
 */
class MdCalendar {
    /**
     * @param {?} _elementRef
     * @param {?} _intl
     * @param {?} _ngZone
     * @param {?} _dateAdapter
     * @param {?} _dateFormats
     * @param {?} changeDetectorRef
     */
    constructor(_elementRef, _intl, _ngZone, _dateAdapter, _dateFormats, changeDetectorRef) {
        this._elementRef = _elementRef;
        this._intl = _intl;
        this._ngZone = _ngZone;
        this._dateAdapter = _dateAdapter;
        this._dateFormats = _dateFormats;
        this._isCompatibilityMode = false;
        /**
         * Whether the calendar should be started in month or year view.
         */
        this.startView = 'month';
        this.type = 'normal';
        /**
         * Emits when the currently selected date changes.
         */
        this.selectedChange = new EventEmitter();
        /**
         * Emits when any date is selected.
         */
        this.userSelection = new EventEmitter();
        this.months = Array.from({ length: 12 }, (/**
         * @param {?} v
         * @param {?} k
         * @return {?}
         */
        (v, k) => k));
        /**
         * Date filter for the month and year views.
         */
        this._dateFilterForViews = (/**
         * @param {?} date
         * @return {?}
         */
        (date) => {
            return !!date && (!this.dateFilter || this.dateFilter(date)) && (!this.minDate || this._dateAdapter.compareDate(date, this.minDate) >= 0) && (!this.maxDate || this._dateAdapter.compareDate(date, this.maxDate) <= 0);
        });
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        if (!this._dateFormats) {
            throw createMissingDateImplError('MD_DATE_FORMATS');
        }
        this._intlChanges = _intl.changes.pipe(merge(this._dateFormats._getChanges)).subscribe((/**
         * @return {?}
         */
        () => {
            changeDetectorRef.markForCheck();
        }));
        this._dateAdapterr = ((/** @type {?} */ (_dateAdapter)));
    }
    /**
     * The current active date. This determines which time period is shown and which date is
     * highlighted when using keyboard navigation.
     * @return {?}
     */
    get _activeNextDate() {
        /** @type {?} */
        const date = this._dateAdapter.clone(this._activeDate);
        if (((/** @type {?} */ (date)))._d) {
            ((/** @type {?} */ (date))).add(1, 'jMonth');
            // (date as any)._d.setMonth((date as any)._d.getMonth() + 1);
            return ((/** @type {?} */ (date)))._d;
        }
        else {
            ((/** @type {?} */ (date))).setMonth(((/** @type {?} */ (date))).getMonth() + 1);
            return date;
        }
    }
    /**
     * @return {?}
     */
    get _activeDate() {
        return this._clampedActiveDate;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set _activeDate(value) {
        this._clampedActiveDate = this._dateAdapter.clampDate(value, this.minDate, this.maxDate);
    }
    /**
     * The label for the current calendar view.
     * @return {?}
     */
    get _periodButtonText() {
        return this._monthView ? this._dateAdapter.format(this._activeDate, this._dateFormats.display.monthYearLabel)
            .toLocaleUpperCase() : this._dateAdapter.getYearName(this._activeDate);
    }
    /**
     * @return {?}
     */
    get _periodButtonLabel() {
        return this._monthView ? this._intl.switchToYearViewLabel : this._intl.switchToMonthViewLabel;
    }
    /**
     * The label for the the previous button.
     * @return {?}
     */
    get _prevButtonLabel() {
        return this._monthView ? this._intl.prevMonthLabel : this._intl.prevYearLabel;
    }
    /**
     * The label for the the next button.
     * @return {?}
     */
    get _nextButtonLabel() {
        return this._monthView ? this._intl.nextMonthLabel : this._intl.nextYearLabel;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._activeDate = this.startAt || this._dateAdapter.today();
        this._focusActiveCell();
        this._monthView = this.startView != 'year';
        //fill year drop
        // main page
        if (this.minDate && this.maxDate) {
            this.foods = Array.from({ length: 100 }, (/**
             * @param {?} v
             * @param {?} k
             * @return {?}
             */
            (v, k) => 10 - k));
        }
        else if (this.minDate && !this.maxDate) {
            this.foods = Array.from({ length: 100 }, (/**
             * @param {?} v
             * @param {?} k
             * @return {?}
             */
            (v, k) => 10 - k));
        }
        else if (this.maxDate && !this.minDate) {
            this.foods = Array.from({ length: 100 }, (/**
             * @param {?} v
             * @param {?} k
             * @return {?}
             */
            (v, k) => 10 - k));
        }
        //for old dates like birthday
        else {
            this.foods = Array.from({ length: 100 }, (/**
             * @param {?} v
             * @param {?} k
             * @return {?}
             */
            (v, k) => 10 - k));
        }
    }
    /**
     * @param {?} v
     * @return {?}
     */
    _showDropYear(v) {
        return this._dateAdapter.getYear(this._dateAdapter.addCalendarYears((this.startAt || (/** @type {?} */ (new Date()))), v));
    }
    /**
     * @param {?} v
     * @return {?}
     */
    _showDropMonth(v) {
        return this._dateAdapter.getMonthNames("long")[v];
    }
    /**
     * @param {?} v
     * @return {?}
     */
    _yearSelected(v) {
        this._activeDate = this._dateAdapter.addCalendarYears(this.startAt || (/** @type {?} */ (new Date())), v.value);
    }
    /**
     * @param {?} v
     * @return {?}
     */
    _monthSelectedd(v) {
        /** @type {?} */
        let daysInMonth = this._dateAdapter.getNumDaysInMonth(this._dateAdapter.createDate(this._dateAdapter.getYear(this._activeDate), v.value, 1));
        this._activeDate = this._dateAdapter.createDate(this._dateAdapter.getYear(this._activeDate), v.value, Math.min(this._dateAdapter.getDate(this._activeDate), daysInMonth));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._intlChanges.unsubscribe();
    }
    /**
     * Handles date selection in the month view.
     * @param {?} date
     * @return {?}
     */
    _dateSelected(date) {
        this.selectedChange.emit(date);
    }
    /**
     * @return {?}
     */
    _userSelected() {
        this.userSelection.emit();
    }
    /**
     * Handles month selection in the year view.
     * @param {?} month
     * @return {?}
     */
    _monthSelected(month) {
        this._activeDate = month;
        this._monthView = true;
    }
    /**
     * Handles user clicks on the period label.
     * @return {?}
     */
    _currentPeriodClicked() {
        this._monthView = !this._monthView;
    }
    /**
     * Handles user clicks on the previous button.
     * @return {?}
     */
    _previousClicked() {
        this._activeDate = this._monthView ? this._dateAdapter.addCalendarMonths(this._activeDate, -1) : this._dateAdapter.addCalendarYears(this._activeDate, -1);
    }
    /**
     * Handles user clicks on the next button.
     * @return {?}
     */
    _nextClicked() {
        this._activeDate = this._monthView ? this._dateAdapter.addCalendarMonths(this._activeDate, 1) : this._dateAdapter.addCalendarYears(this._activeDate, 1);
    }
    /**
     * Whether the previous period button is enabled.
     * @return {?}
     */
    _previousEnabled() {
        if (!this.minDate) {
            return true;
        }
        return !this.minDate || !this._isSameView(this._activeDate, this.minDate);
    }
    /**
     * Whether the next period button is enabled.
     * @return {?}
     */
    _nextEnabled() {
        return !this.maxDate || !this._isSameView(this._activeDate, this.maxDate);
    }
    /**
     * Handles keydown events on the calendar body.
     * @param {?} event
     * @return {?}
     */
    _handleCalendarBodyKeydown(event) {
        // TODO(mmalerba): We currently allow keyboard navigation to disabled dates, but just prevent
        // disabled ones from being selected. This may not be ideal, we should look into whether
        // navigation should skip over disabled dates, and if so, how to implement that efficiently.
        if (this._monthView) {
            this._handleCalendarBodyKeydownInMonthView(event);
        }
        else {
            this._handleCalendarBodyKeydownInYearView(event);
        }
    }
    /**
     * Focuses the active cell after the microtask queue is empty.
     * @return {?}
     */
    _focusActiveCell() {
        this._ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => first.call(this._ngZone.onStable).subscribe((/**
         * @return {?}
         */
        () => {
            this._elementRef.nativeElement.querySelector('.mat-calendar-body-active').focus();
        }))));
    }
    /**
     * Whether the two dates represent the same view in the current view mode (month or year).
     * @private
     * @param {?} date1
     * @param {?} date2
     * @return {?}
     */
    _isSameView(date1, date2) {
        return this._monthView ? this._dateAdapter.getYear(date1) == this._dateAdapter.getYear(date2) && this._dateAdapter.getMonth(date1) == this._dateAdapter.getMonth(date2) : this._dateAdapter.getYear(date1) == this._dateAdapter.getYear(date2);
    }
    /**
     * Handles keydown events on the calendar body when calendar is in month view.
     * @private
     * @param {?} event
     * @return {?}
     */
    _handleCalendarBodyKeydownInMonthView(event) {
        switch (event.keyCode) {
            case LEFT_ARROW:
                this._activeDate = this._dateAdapter.addCalendarDays(this._activeDate, -1);
                break;
            case RIGHT_ARROW:
                this._activeDate = this._dateAdapter.addCalendarDays(this._activeDate, 1);
                break;
            case UP_ARROW:
                this._activeDate = this._dateAdapter.addCalendarDays(this._activeDate, -7);
                break;
            case DOWN_ARROW:
                this._activeDate = this._dateAdapter.addCalendarDays(this._activeDate, 7);
                break;
            case HOME:
                this._activeDate = this._dateAdapter.addCalendarDays(this._activeDate, 1 - this._dateAdapter.getDate(this._activeDate));
                break;
            case END:
                this._activeDate = this._dateAdapter.addCalendarDays(this._activeDate, (this._dateAdapter.getNumDaysInMonth(this._activeDate) - this._dateAdapter.getDate(this._activeDate)));
                break;
            case PAGE_UP:
                this._activeDate = event.altKey ? this._dateAdapter.addCalendarYears(this._activeDate, -1) : this._dateAdapter.addCalendarMonths(this._activeDate, -1);
                break;
            case PAGE_DOWN:
                this._activeDate = event.altKey ? this._dateAdapter.addCalendarYears(this._activeDate, 1) : this._dateAdapter.addCalendarMonths(this._activeDate, 1);
                break;
            case ENTER:
                if (this._dateFilterForViews(this._activeDate)) {
                    this._dateSelected(this._activeDate);
                    // Prevent unexpected default actions such as form submission.
                    event.preventDefault();
                }
                return;
            default:
                // Don't prevent default or focus active cell on keys that we don't explicitly handle.
                return;
        }
        this._focusActiveCell();
        // Prevent unexpected default actions such as form submission.
        event.preventDefault();
    }
    /**
     * Handles keydown events on the calendar body when calendar is in year view.
     * @private
     * @param {?} event
     * @return {?}
     */
    _handleCalendarBodyKeydownInYearView(event) {
        switch (event.keyCode) {
            case LEFT_ARROW:
                this._activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, -1);
                break;
            case RIGHT_ARROW:
                this._activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, 1);
                break;
            case UP_ARROW:
                this._activeDate = this._prevMonthInSameCol(this._activeDate);
                break;
            case DOWN_ARROW:
                this._activeDate = this._nextMonthInSameCol(this._activeDate);
                break;
            case HOME:
                this._activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, -this._dateAdapter.getMonth(this._activeDate));
                break;
            case END:
                this._activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, 11 - this._dateAdapter.getMonth(this._activeDate));
                break;
            case PAGE_UP:
                this._activeDate = this._dateAdapter.addCalendarYears(this._activeDate, event.altKey ? -10 : -1);
                break;
            case PAGE_DOWN:
                this._activeDate = this._dateAdapter.addCalendarYears(this._activeDate, event.altKey ? 10 : 1);
                break;
            case ENTER:
                this._monthSelected(this._activeDate);
                break;
            default:
                // Don't prevent default or focus active cell on keys that we don't explicitly handle.
                return;
        }
        this._focusActiveCell();
        // Prevent unexpected default actions such as form submission.
        event.preventDefault();
    }
    /**
     * Determine the date for the month that comes before the given month in the same column in the
     * calendar table.
     * @private
     * @param {?} date
     * @return {?}
     */
    _prevMonthInSameCol(date) {
        // Determine how many months to jump forward given that there are 2 empty slots at the beginning
        // of each year.
        /** @type {?} */
        const increment = this._dateAdapter.getMonth(date) <= 4 ? -5 : (this._dateAdapter.getMonth(date) >= 7 ? -7 : -12);
        return this._dateAdapter.addCalendarMonths(date, increment);
    }
    /**
     * Determine the date for the month that comes after the given month in the same column in the
     * calendar table.
     * @private
     * @param {?} date
     * @return {?}
     */
    _nextMonthInSameCol(date) {
        // Determine how many months to jump forward given that there are 2 empty slots at the beginning
        // of each year.
        /** @type {?} */
        const increment = this._dateAdapter.getMonth(date) <= 4 ? 7 : (this._dateAdapter.getMonth(date) >= 7 ? 5 : 12);
        return this._dateAdapter.addCalendarMonths(date, increment);
    }
}
MdCalendar.decorators = [
    { type: Component, args: [{
                selector: 'md-calendar, mat-calendar',
                template: "<div class=\"calendar-drops\">\r\n    <!--<button *ngIf=\"!_isCompatibilityMode\" [disableRipple]=\"true\" mat-button class=\"mat-calendar-period-button\"-->\r\n    <!--(click)=\"_currentPeriodClicked()\" [attr.aria-label]=\"_periodButtonLabel\">-->\r\n    <!--\u0645\u0627\u0647-->\r\n    <!--<div class=\"mat-calendar-arrow\" [class.mat-calendar-invert]=\"!_monthView\"></div>-->\r\n    <!--</button>-->\r\n    <!--<button *ngIf=\"_isCompatibilityMode\" mat-button class=\"mat-calendar-period-button\"-->\r\n    <!--(click)=\"_currentPeriodClicked()\" [attr.aria-label]=\"_periodButtonLabel\">-->\r\n    <!--\u0645\u0627\u0647-->\r\n    <!--<div class=\"mat-calendar-arrow\" [class.mat-calendar-invert]=\"!_monthView\"></div>-->\r\n    <!--</button>-->\r\n\r\n    <mat-select #yearsDrop placeholder=\"\u0633\u0627\u0644\" style=\"width: 50px;\"\r\n                (change)=\"yearsDrop.value = null;_yearSelected($event)\"\r\n                panelClass=\"drops-panels years {{(_dateAdapterr?.lang == 'en') ? 'latin' :''}}\">\r\n        <mat-option *ngFor=\"let food of foods\" [value]=\"food\">\r\n            {{ _showDropYear(food) }}\r\n        </mat-option>\r\n    </mat-select>\r\n    <mat-select #monthsDrop placeholder=\"\u0645\u0627\u0647\" style=\"width: 70px;\"\r\n                (change)=\"monthsDrop.value = null;_monthSelectedd($event)\"\r\n                panelClass=\"drops-panels months {{(_dateAdapterr?.lang == 'en') ? 'latin' :''}}\">\r\n        <mat-option *ngFor=\"let month of months\" [value]=\"month\">\r\n            {{ _showDropMonth(month) }}\r\n        </mat-option>\r\n    </mat-select>\r\n\r\n</div>\r\n<div class=\"mat-calendar-header\">\r\n    <div class=\"mat-calendar-controls\" [class.month]=\"_monthView\">\r\n        <!--\r\n          TODO(mmalerba): Clean up duplicated compatibility mode code when we have a better way to do\r\n          this.\r\n        -->\r\n\r\n        <!--additions by puria-->\r\n        <ng-content></ng-content>\r\n\r\n\r\n        <!-- Check for compatibility mode and use correct prefix for md-icon-button. -->\r\n        <button *ngIf=\"!_isCompatibilityMode\" mat-icon-button class=\"mat-calendar-previous-button\"\r\n                [disabled]=\"!_previousEnabled()\" (click)=\"_previousClicked()\"\r\n                [attr.aria-label]=\"_prevButtonLabel\">\r\n        </button>\r\n        <button *ngIf=\"_isCompatibilityMode\" mat-icon-button class=\"mat-calendar-previous-button\"\r\n                [disabled]=\"!_previousEnabled()\" (click)=\"_previousClicked()\"\r\n                [attr.aria-label]=\"_prevButtonLabel\">\r\n        </button>\r\n\r\n        <!-- Check for compatibility mode and use correct prefix for md-button. -->\r\n\r\n\r\n        <div class=\"mat-calendar-spacer\"></div>\r\n\r\n\r\n        <!-- Check for compatibility mode and use correct prefix for md-icon-button. -->\r\n        <button *ngIf=\"!_isCompatibilityMode\" mat-icon-button class=\"mat-calendar-next-button\"\r\n                [disabled]=\"!_nextEnabled()\" (click)=\"_nextClicked()\"\r\n                [attr.aria-label]=\"_nextButtonLabel\">\r\n        </button>\r\n        <button *ngIf=\"_isCompatibilityMode\" mat-icon-button class=\"mat-calendar-next-button\"\r\n                [disabled]=\"!_nextEnabled()\" (click)=\"_nextClicked()\"\r\n                [attr.aria-label]=\"_nextButtonLabel\">\r\n        </button>\r\n    </div>\r\n</div>\r\n\r\n<div class=\"mat-calendar-content\" (keydown)=\"_handleCalendarBodyKeydown($event)\"\r\n     [ngSwitch]=\"_monthView\" cdkMonitorSubtreeFocus>\r\n    <md-month-view\r\n        *ngSwitchCase=\"true\"\r\n        [activeDate]=\"_activeDate\"\r\n        [subedDate]=\"subed\"\r\n        [selected]=\"selected\"\r\n        [dateFilter]=\"_dateFilterForViews\"\r\n        (selectedChange)=\"_dateSelected($event)\"\r\n        (userSelection)=\"_userSelected()\">\r\n    </md-month-view>\r\n    <md-month-view\r\n        *ngIf=\"type == 'wide' && _monthView\"\r\n        [activeDate]=\"_activeNextDate\"\r\n        [subedDate]=\"subed\"\r\n        [selected]=\"selected\"\r\n        [dateFilter]=\"_dateFilterForViews\"\r\n        (selectedChange)=\"_dateSelected($event)\"\r\n        (userSelection)=\"_userSelected()\">\r\n    </md-month-view>\r\n    <md-year-view\r\n        *ngSwitchDefault\r\n        [activeDate]=\"_activeDate\"\r\n        [selected]=\"selected\"\r\n        [dateFilter]=\"_dateFilterForViews\"\r\n        (selectedChange)=\"_monthSelected($event)\">\r\n    </md-year-view>\r\n</div>\r\n",
                host: {
                    'class': 'mat-calendar', '[class.wide-calendar]': '(type == "wide")'
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".mat-calendar{display:block}.mat-calendar-header{padding:16px}.mat-calendar-content{margin:0;padding:0 8px 8px;outline:0}.mat-calendar-controls{display:flex;margin:0}.mat-calendar-spacer{flex:1 1 auto}.mat-calendar-period-button{min-width:0}.mat-calendar-period-button:hover .mat-button-focus-overlay{opacity:0}.subed{background:#ffd2de}.calendar-drops{position:absolute;top:12px;left:26px;display:flex;flex-direction:row-reverse;align-items:flex-end}.calendar-drops .mat-select-value{text-align:left}.drops-panels.months .mat-select-panel{width:70px}.drops-panels .mat-select-panel{width:50px}.drops-panels .mat-option-text{font-size:13px}.mat-calendar-arrow{display:inline-block;width:0;height:0;border-left:5px solid transparent;border-right:5px solid transparent;border-top-width:5px;border-top-style:solid;margin:0!important;vertical-align:middle}.mat-calendar-arrow.mat-calendar-invert{transform:rotate(180deg)}[dir=rtl] .mat-calendar-arrow{margin:0 5px 0 0}.mat-calendar-next-button,.mat-calendar-previous-button{position:relative}.mat-calendar-next-button::after,.mat-calendar-previous-button::after{top:0;left:0;right:0;bottom:0;position:absolute;content:\"\";margin:15.5px;border:0 solid currentColor;border-top-width:2px}[dir=rtl] .mat-calendar-next-button,[dir=rtl] .mat-calendar-previous-button{transform:rotate(180deg)}.mat-calendar-previous-button::after{border-left-width:2px;transform:translateX(2px) rotate(-45deg) scale(-1)}.mat-calendar-next-button::after{border-right-width:2px;transform:translateX(-2px) rotate(45deg) scale(-1)}.mat-calendar-table{border-spacing:0;border-collapse:collapse;width:100%}.mat-calendar-table-header th{text-align:center;padding:0 0 8px}.mat-calendar-table-header-divider{position:relative;height:1px}.mat-calendar-table-header-divider::after{content:\"\";position:absolute;top:0;left:-8px;right:-8px;height:1px}"]
            }] }
];
/** @nocollapse */
MdCalendar.ctorParameters = () => [
    { type: ElementRef },
    { type: MdDatepickerIntl },
    { type: NgZone },
    { type: DateAdapter, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_FORMATS,] }] },
    { type: ChangeDetectorRef }
];
MdCalendar.propDecorators = {
    startAt: [{ type: Input }],
    startView: [{ type: Input }],
    type: [{ type: Input }],
    selected: [{ type: Input }],
    subed: [{ type: Input }],
    minDate: [{ type: Input }],
    maxDate: [{ type: Input }],
    dateFilter: [{ type: Input }],
    selectedChange: [{ type: Output }],
    userSelection: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._intlChanges;
    /** @type {?} */
    MdCalendar.prototype._isCompatibilityMode;
    /**
     * A date representing the period (month or year) to start the calendar in.
     * @type {?}
     */
    MdCalendar.prototype.startAt;
    /**
     * Whether the calendar should be started in month or year view.
     * @type {?}
     */
    MdCalendar.prototype.startView;
    /** @type {?} */
    MdCalendar.prototype.type;
    /**
     * The currently selected date.
     * @type {?}
     */
    MdCalendar.prototype.selected;
    /** @type {?} */
    MdCalendar.prototype.subed;
    /**
     * The minimum selectable date.
     * @type {?}
     */
    MdCalendar.prototype.minDate;
    /**
     * The maximum selectable date.
     * @type {?}
     */
    MdCalendar.prototype.maxDate;
    /**
     * A function used to filter which dates are selectable.
     * @type {?}
     */
    MdCalendar.prototype.dateFilter;
    /**
     * Emits when the currently selected date changes.
     * @type {?}
     */
    MdCalendar.prototype.selectedChange;
    /**
     * Emits when any date is selected.
     * @type {?}
     */
    MdCalendar.prototype.userSelection;
    /** @type {?} */
    MdCalendar.prototype.months;
    /**
     * Date filter for the month and year views.
     * @type {?}
     */
    MdCalendar.prototype._dateFilterForViews;
    /** @type {?} */
    MdCalendar.prototype._dateAdapterr;
    /** @type {?} */
    MdCalendar.prototype.foods;
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._clampedActiveDate;
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._clampedSubedDate;
    /**
     * Whether the calendar is in month view.
     * @type {?}
     */
    MdCalendar.prototype._monthView;
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._intl;
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._ngZone;
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdCalendar.prototype._dateFormats;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/datepicker.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Used to generate a unique ID for each datepicker instance.
 * @type {?}
 */
let datepickerUid = 0;
/**
 * Injection token that determines the scroll handling while the calendar is open.
 * @type {?}
 */
const MD_DATEPICKER_SCROLL_STRATEGY = new InjectionToken('md-datepicker-scroll-strategy');
/**
 * \@docs-private
 * @param {?} overlay
 * @return {?}
 */
function MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER_FACTORY(overlay) {
    return (/**
     * @return {?}
     */
    () => overlay.scrollStrategies.reposition());
}
/**
 * \@docs-private
 * @type {?}
 */
const MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER = {
    provide: MD_DATEPICKER_SCROLL_STRATEGY, deps: [Overlay], useFactory: MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER_FACTORY,
};
/**
 * Component used as the content for the datepicker dialog and popup. We use this instead of using
 * MdCalendar directly as the content so we can control the initial focus. This also gives us a
 * place to put additional features of the popup that are not part of the calendar itself in the
 * future. (e.g. confirmation buttons).
 * \@docs-private
 * @template D
 */
class MdDatepickerContent {
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._calendar._focusActiveCell();
    }
    /**
     * Handles keydown event on datepicker content.
     * @param {?} event The event.
     * @return {?}
     */
    _handleKeydown(event) {
        if (event.keyCode === ESCAPE) {
            this.datepicker.close();
            event.preventDefault();
            event.stopPropagation();
        }
    }
}
MdDatepickerContent.decorators = [
    { type: Component, args: [{
                selector: 'md-datepicker-content, mat-datepicker-content',
                template: "<md-locale-changer>miladi</md-locale-changer>\r\n<md-calendar cdkTrapFocus\r\n             [id]=\"datepicker.id\"\r\n             [type]=\"datepicker.type\"\r\n             [startAt]=\"datepicker.startAt\"\r\n             [startView]=\"datepicker.startView\"\r\n             [minDate]=\"datepicker._minDate\"\r\n             [maxDate]=\"datepicker._maxDate\"\r\n             [dateFilter]=\"datepicker._dateFilter\"\r\n             [selected]=\"datepicker._selected\"\r\n             [subed]=\"datepicker.subed\"\r\n             (selectedChange)=\"datepicker._select($event)\"\r\n             (userSelection)=\"datepicker.close()\">\r\n</md-calendar>\r\n",
                host: {
                    'class': 'mat-datepicker-content',
                    '[class.mat-datepicker-content-touch]': 'datepicker.touchUi',
                    '(keydown)': '_handleKeydown($event)',
                    '[class.gregorian]': 'datepicker.adapter.lang == "en"',
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".gregorian md-calendar * .superSmall{font-family:vazir,sans-serif!important}.mat-calendar-table-header{line-height:14px}.mlabel{text-align:center;margin-top:-40px;margin-bottom:10px;color:rgba(0,0,0,.54);line-height:10px;pointer-events:none}.mlabel span{font-size:10px;color:#989898;padding:0 2px}.mat-datepicker-content{box-shadow:0 5px 5px -3px rgba(0,0,0,.2),0 8px 10px 1px rgba(0,0,0,.14),0 3px 14px 2px rgba(0,0,0,.12);direction:rtl;display:block}md-locale-changer{display:block}md-locale-changer>*{margin:20px}md-locale-changer mat-button-toggle-group{margin:16px 16px 0}.mat-calendar{width:296px;height:auto}.mat-datepicker-content-touch{box-shadow:0 0 0 0 rgba(0,0,0,.2),0 0 0 0 rgba(0,0,0,.14),0 0 0 0 rgba(0,0,0,.12);display:block;overflow:auto;margin:-24px}.mat-datepicker-content-touch .mat-calendar{min-width:250px;min-height:312px;max-width:750px;max-height:788px}@media all and (orientation:landscape){.mat-datepicker-content-touch .mat-calendar{width:64vh;height:80vh}}@media all and (orientation:portrait){.mat-datepicker-content-touch .mat-calendar{width:80vw;height:100vw}}"]
            }] }
];
MdDatepickerContent.propDecorators = {
    _calendar: [{ type: ViewChild, args: [MdCalendar, { static: true },] }]
};
if (false) {
    /** @type {?} */
    MdDatepickerContent.prototype.datepicker;
    /** @type {?} */
    MdDatepickerContent.prototype._calendar;
}
// TODO(mmalerba): We use a component instead of a directive here so the user can use implicit
// template reference variables (e.g. #d vs #d="mdDatepicker"). We can change this to a directive if
// angular adds support for `exportAs: '$implicit'` on directives.
/**
 * Component responsible for managing the datepicker popup/dialog.
 * @template D
 */
class MdDatepicker {
    /**
     * @param {?} _dialog
     * @param {?} _overlay
     * @param {?} _ngZone
     * @param {?} _viewContainerRef
     * @param {?} _scrollStrategy
     * @param {?} _dateAdapter
     * @param {?} _dir
     * @param {?} _document
     */
    constructor(_dialog, _overlay, _ngZone, _viewContainerRef, _scrollStrategy, _dateAdapter, _dir, _document) {
        this._dialog = _dialog;
        this._overlay = _overlay;
        this._ngZone = _ngZone;
        this._viewContainerRef = _viewContainerRef;
        this._scrollStrategy = _scrollStrategy;
        this._dateAdapter = _dateAdapter;
        this._dir = _dir;
        this._document = _document;
        /**
         * The view that the calendar should start in.
         */
        this.startView = 'month';
        this.type = 'normal';
        /**
         * Whether the calendar UI is in touch mode. In touch mode the calendar opens in a dialog rather
         * than a popup and elements have more padding to allow for bigger touch targets.
         */
        this.touchUi = false;
        /**
         * Emits new selected date when selected date changes.
         * @deprecated Switch to the `dateChange` and `dateInput` binding on the input element.
         */
        this.selectedChanged = new EventEmitter();
        /**
         * Whether the calendar is open.
         */
        this.opened = false;
        /**
         * The id for the datepicker calendar.
         */
        this.id = `md-datepicker-${datepickerUid++}`;
        this._validSelected = null;
        this._validSubed = null;
        /**
         * The element that was focused before the datepicker was opened.
         */
        this._focusedElementBeforeOpen = null;
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        this._dateAdapter;
    }
    //observable
    /**
     * The currently selected date.
     * @return {?}
     */
    get subed() {
        return this._validSubed;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set subed(value) {
        this._validSubed = value;
    }
    /**
     * The date to open the calendar to initially.
     * @return {?}
     */
    get startAt() {
        // If an explicit startAt is set we start there, otherwise we start at whatever the currently
        // selected value is.
        return this._startAt || (this._datepickerInput ? this._datepickerInput.value : null);
    }
    /**
     * @param {?} date
     * @return {?}
     */
    set startAt(date) {
        this._startAt = date;
    }
    /**
     * Whether the datepicker pop-up should be disabled.
     * @return {?}
     */
    get disabled() {
        return this._disabled === undefined ? this._datepickerInput.disabled : this._disabled;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this._disabled = coerceBooleanProperty(value);
    }
    /**
     * The currently selected date.
     * @return {?}
     */
    get _selected() {
        return this._validSelected;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set _selected(value) {
        this._validSelected = value;
    }
    /**
     * The minimum selectable date.
     * @return {?}
     */
    get _minDate() {
        return this._datepickerInput && this._datepickerInput.min;
    }
    /**
     * The maximum selectable date.
     * @return {?}
     */
    get _maxDate() {
        return this._datepickerInput && this._datepickerInput.max;
    }
    /**
     * @return {?}
     */
    get _dateFilter() {
        return this._datepickerInput && this._datepickerInput._dateFilter;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.close();
        if (this._popupRef) {
            this._popupRef.dispose();
        }
        if (this._inputSubscription) {
            this._inputSubscription.unsubscribe();
        }
    }
    /**
     * Selects the given date
     * @param {?} date
     * @return {?}
     */
    _select(date) {
        /** @type {?} */
        const oldValue = this._selected;
        if (((/** @type {?} */ (date)))._d) {
            this._selected = ((/** @type {?} */ (date)))._d;
        }
        else {
            this._selected = date;
        }
        this.selectedChanged.emit(this._selected);
    }
    /**
     * Register an input with this datepicker.
     * @param {?} input The datepicker input to register with this datepicker.
     * @return {?}
     */
    _registerInput(input) {
        if (this._datepickerInput) {
            throw Error('An MdDatepicker can only be associated with a single input.');
        }
        this._datepickerInput = input;
        this._inputSubscription = this._datepickerInput._valueChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        (value) => this._selected = value));
    }
    /**
     * Open the calendar.
     * @return {?}
     */
    open() {
        if (this.opened || this.disabled) {
            return;
        }
        if (!this._datepickerInput) {
            throw Error('Attempted to open an MdDatepicker with no associated input.');
        }
        if (this._document) {
            this._focusedElementBeforeOpen = this._document.activeElement;
        }
        this.touchUi ? this._openAsDialog() : this._openAsPopup();
        this.opened = true;
    }
    /**
     * Close the calendar.
     * @return {?}
     */
    close() {
        if (!this.opened) {
            return;
        }
        if (this._popupRef && this._popupRef.hasAttached()) {
            this._popupRef.detach();
        }
        if (this._dialogRef) {
            this._dialogRef.close();
            this._dialogRef = null;
        }
        if (this._calendarPortal && this._calendarPortal.isAttached) {
            this._calendarPortal.detach();
        }
        if (this._focusedElementBeforeOpen && typeof this._focusedElementBeforeOpen.focus === 'function') {
            this._focusedElementBeforeOpen.focus();
            this._focusedElementBeforeOpen = null;
        }
        this.opened = false;
    }
    /**
     * Open the calendar as a dialog.
     * @private
     * @return {?}
     */
    _openAsDialog() {
        this._dialogRef = this._dialog.open(MdDatepickerContent, {
            direction: this._dir ? this._dir.value : 'ltr', viewContainerRef: this._viewContainerRef,
        });
        this._dialogRef.afterClosed().subscribe((/**
         * @return {?}
         */
        () => this.close()));
        this._dialogRef.componentInstance.datepicker = this;
    }
    /**
     * Open the calendar as a popup.
     * @private
     * @return {?}
     */
    _openAsPopup() {
        if (!this._calendarPortal) {
            this._calendarPortal = (/** @type {?} */ (new ComponentPortal(MdDatepickerContent, this._viewContainerRef)));
        }
        if (!this._popupRef) {
            this._createPopup();
        }
        if (!this._popupRef.hasAttached()) {
            /** @type {?} */
            const componentRef = (/** @type {?} */ (this._popupRef.attach(this._calendarPortal)));
            componentRef.instance.datepicker = (/** @type {?} */ (this));
            // Update the position once the calendar has rendered.
            first.call(this._ngZone.onStable).subscribe((/**
             * @return {?}
             */
            () => this._popupRef.updatePosition()));
        }
        this._popupRef.backdropClick().subscribe((/**
         * @return {?}
         */
        () => this.close()));
    }
    /**
     * Create the popup.
     * @private
     * @return {?}
     */
    _createPopup() {
        /** @type {?} */
        const overlayState = new OverlayConfig();
        overlayState.positionStrategy = this._createPopupPositionStrategy();
        overlayState.hasBackdrop = true;
        overlayState.backdropClass = 'md-overlay-transparent-backdrop';
        overlayState.direction = this._dir ? this._dir.value : 'ltr';
        overlayState.scrollStrategy = this._scrollStrategy();
        this._popupRef = this._overlay.create(overlayState);
    }
    /**
     * Create the popup PositionStrategy.
     * @private
     * @return {?}
     */
    _createPopupPositionStrategy() {
        return this._overlay.position()
            .connectedTo(this._datepickerInput.getPopupConnectionElementRef(), {
            originX: 'start', originY: 'bottom'
        }, { overlayX: 'start', overlayY: 'top' })
            .withFallbackPosition({ originX: 'start', originY: 'top' }, { overlayX: 'start', overlayY: 'bottom' })
            .withFallbackPosition({ originX: 'end', originY: 'bottom' }, { overlayX: 'end', overlayY: 'top' })
            .withFallbackPosition({ originX: 'end', originY: 'top' }, { overlayX: 'end', overlayY: 'bottom' });
    }
}
MdDatepicker.decorators = [
    { type: Component, args: [{
                selector: 'md-datepicker, mat-datepicker',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
MdDatepicker.ctorParameters = () => [
    { type: MatDialog },
    { type: Overlay },
    { type: NgZone },
    { type: ViewContainerRef },
    { type: undefined, decorators: [{ type: Inject, args: [MD_DATEPICKER_SCROLL_STRATEGY,] }] },
    { type: DateAdapter, decorators: [{ type: Optional }] },
    { type: Directionality, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [DOCUMENT,] }] }
];
MdDatepicker.propDecorators = {
    subed: [{ type: Input }],
    startAt: [{ type: Input }],
    startView: [{ type: Input }],
    type: [{ type: Input }],
    touchUi: [{ type: Input }],
    disabled: [{ type: Input }],
    selectedChanged: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._startAt;
    /**
     * The view that the calendar should start in.
     * @type {?}
     */
    MdDatepicker.prototype.startView;
    /** @type {?} */
    MdDatepicker.prototype.type;
    /**
     * Whether the calendar UI is in touch mode. In touch mode the calendar opens in a dialog rather
     * than a popup and elements have more padding to allow for bigger touch targets.
     * @type {?}
     */
    MdDatepicker.prototype.touchUi;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._disabled;
    /**
     * Emits new selected date when selected date changes.
     * @deprecated Switch to the `dateChange` and `dateInput` binding on the input element.
     * @type {?}
     */
    MdDatepicker.prototype.selectedChanged;
    /**
     * Whether the calendar is open.
     * @type {?}
     */
    MdDatepicker.prototype.opened;
    /**
     * The id for the datepicker calendar.
     * @type {?}
     */
    MdDatepicker.prototype.id;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._validSelected;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._validSubed;
    /**
     * A reference to the overlay when the calendar is opened as a popup.
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._popupRef;
    /**
     * A reference to the dialog when the calendar is opened as a dialog.
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._dialogRef;
    /**
     * A portal containing the calendar for this datepicker.
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._calendarPortal;
    /**
     * The input element this datepicker is associated with.
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._datepickerInput;
    /**
     * The element that was focused before the datepicker was opened.
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._focusedElementBeforeOpen;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._inputSubscription;
    /** @type {?} */
    MdDatepicker.prototype.adapter;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._dialog;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._overlay;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._ngZone;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._scrollStrategy;
    /** @type {?} */
    MdDatepicker.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._dir;
    /**
     * @type {?}
     * @private
     */
    MdDatepicker.prototype._document;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/datepicker-input.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MD_DATEPICKER_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => MdDatepickerInput)),
    multi: true
};
/** @type {?} */
const MD_DATEPICKER_VALIDATORS = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => MdDatepickerInput)),
    multi: true
};
/**
 * An event used for datepicker input and change events. We don't always have access to a native
 * input or change event because the event may have been triggered by the user clicking on the
 * calendar popup. For consistency, we always use MdDatepickerInputEvent instead.
 * @template D
 */
class MdDatepickerInputEvent {
    /**
     * @param {?} target
     * @param {?} targetElement
     */
    constructor(target, targetElement) {
        this.target = target;
        this.targetElement = targetElement;
        this.value = this.target.value;
    }
}
if (false) {
    /**
     * The new value for the target datepicker input.
     * @type {?}
     */
    MdDatepickerInputEvent.prototype.value;
    /** @type {?} */
    MdDatepickerInputEvent.prototype.target;
    /** @type {?} */
    MdDatepickerInputEvent.prototype.targetElement;
}
/**
 * Directive used to connect an input to a MdDatepicker.
 * @template D
 */
class MdDatepickerInput {
    /**
     * @param {?} _elementRef
     * @param {?} _renderer
     * @param {?} _dateAdapter
     * @param {?} _dateFormats
     * @param {?} _mdFormField
     */
    constructor(_elementRef, _renderer, _dateAdapter, _dateFormats, _mdFormField) {
        this._elementRef = _elementRef;
        this._renderer = _renderer;
        this._dateAdapter = _dateAdapter;
        this._dateFormats = _dateFormats;
        this._mdFormField = _mdFormField;
        /**
         * Emits when a `change` event is fired on this `<input>`.
         */
        this.dateChange = new EventEmitter();
        /**
         * Emits when an `input` event is fired on this `<input>`.
         */
        this.dateInput = new EventEmitter();
        /**
         * Emits when the value changes (either due to user input or programmatic change).
         */
        this._valueChange = new EventEmitter();
        this._onTouched = (/**
         * @return {?}
         */
        () => {
        });
        this._cvaOnChange = (/**
         * @return {?}
         */
        () => {
        });
        this._validatorOnChange = (/**
         * @return {?}
         */
        () => {
        });
        /**
         * The form control validator for whether the input parses.
         */
        this._parseValidator = (/**
         * @return {?}
         */
        () => {
            return this._lastValueValid ?
                null : { 'mdDatepickerParse': { 'text': this._elementRef.nativeElement.value } };
        });
        /**
         * The form control validator for the min date.
         */
        this._minValidator = (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            return (!this.min || !control.value ||
                this._dateAdapter.compareDate(this.min, control.value) <= 0) ?
                null : { 'mdDatepickerMin': { 'min': this.min, 'actual': control.value } };
        });
        /**
         * The form control validator for the max date.
         */
        this._maxValidator = (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            return (!this.max || !control.value ||
                this._dateAdapter.compareDate(this.max, control.value) >= 0) ?
                null : { 'mdDatepickerMax': { 'max': this.max, 'actual': control.value } };
        });
        /**
         * The form control validator for the date filter.
         */
        this._filterValidator = (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            return !this._dateFilter || !control.value || this._dateFilter(control.value) ?
                null : { 'mdDatepickerFilter': true };
        });
        /**
         * The combined form control validator for this input.
         */
        this._validator = Validators.compose([this._parseValidator, this._minValidator, this._maxValidator, this._filterValidator]);
        /**
         * Whether the last value set on the input was valid.
         */
        this._lastValueValid = false;
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        if (!this._dateFormats) {
            throw createMissingDateImplError('MD_DATE_FORMATS');
        }
    }
    /**
     * The datepicker that this input is associated with.
     * @param {?} value
     * @return {?}
     */
    set mdDatepicker(value) {
        if (value) {
            this._datepicker = value;
            this._datepicker._registerInput(this);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set matDatepicker(value) {
        this.mdDatepicker = value;
    }
    /**
     * @param {?} filter
     * @return {?}
     */
    set mdDatepickerFilter(filter) {
        this._dateFilter = filter;
        this._validatorOnChange();
    }
    /**
     * @param {?} filter
     * @return {?}
     */
    set matDatepickerFilter(filter) {
        this.mdDatepickerFilter = filter;
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        if (value != null && !this._dateAdapter.isDateInstance(value)) {
            throw Error('Datepicker: value not recognized as a date object by DateAdapter.');
        }
        this._lastValueValid = !value || this._dateAdapter.isValid(value);
        value = this._getValidDateOrNull(value);
        /** @type {?} */
        const oldDate = this.value;
        this._renderer.setProperty(this._elementRef.nativeElement, 'value', value ? this._dateAdapter.format(value, this._dateFormats.display.dateInput) : '');
        if (!this._dateAdapter.sameDate(oldDate, value)) {
            this._value = value;
            this._valueChange.emit(value);
        }
    }
    /**
     * The minimum valid date.
     * @return {?}
     */
    get min() {
        return this._min;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set min(value) {
        this._min = value;
        this._validatorOnChange();
    }
    /**
     * The maximum valid date.
     * @return {?}
     */
    get max() {
        return this._max;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set max(value) {
        this._max = value;
        this._validatorOnChange();
    }
    /**
     * Whether the datepicker-input is disabled.
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this._disabled = coerceBooleanProperty(value);
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        if (this._datepicker) {
            this._datepickerSubscription =
                this._datepicker.selectedChanged.subscribe((/**
                 * @param {?} selected
                 * @return {?}
                 */
                (selected) => {
                    this.value = selected;
                    this._cvaOnChange(selected);
                    this._onTouched();
                    this.dateInput.emit(new MdDatepickerInputEvent(this, this._elementRef.nativeElement));
                    this.dateChange.emit(new MdDatepickerInputEvent(this, this._elementRef.nativeElement));
                }));
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._datepickerSubscription) {
            this._datepickerSubscription.unsubscribe();
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnValidatorChange(fn) {
        this._validatorOnChange = fn;
    }
    /**
     * @param {?} c
     * @return {?}
     */
    validate(c) {
        return this._validator ? this._validator(c) : null;
    }
    /**
     * Gets the element that the datepicker popup should be connected to.
     * @return {?} The element to connect the popup to.
     */
    getPopupConnectionElementRef() {
        return this._mdFormField ? this._mdFormField.underlineRef : this._elementRef;
    }
    // Implemented as part of ControlValueAccessor
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.value = value;
    }
    // Implemented as part of ControlValueAccessor
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this._cvaOnChange = fn;
    }
    // Implemented as part of ControlValueAccessor
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    // Implemented as part of ControlValueAccessor
    /**
     * @param {?} disabled
     * @return {?}
     */
    setDisabledState(disabled) {
        this._renderer.setProperty(this._elementRef.nativeElement, 'disabled', disabled);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _onKeydown(event) {
        if (event.altKey && event.keyCode === DOWN_ARROW) {
            this._datepicker.open();
            event.preventDefault();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    _onInput(value) {
        /** @type {?} */
        let date = this._dateAdapter.parse(value, this._dateFormats.parse.dateInput);
        this._lastValueValid = !date || this._dateAdapter.isValid(date);
        date = this._getValidDateOrNull(date);
        this._cvaOnChange(date);
        this._valueChange.emit(date);
        this.dateInput.emit(new MdDatepickerInputEvent(this, this._elementRef.nativeElement));
    }
    /**
     * @return {?}
     */
    _onChange() {
        this.dateChange.emit(new MdDatepickerInputEvent(this, this._elementRef.nativeElement));
    }
    /**
     * @private
     * @param {?} obj The object to check.
     * @return {?} The given object if it is both a date instance and valid, otherwise null.
     */
    _getValidDateOrNull(obj) {
        return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
    }
}
MdDatepickerInput.decorators = [
    { type: Directive, args: [{
                selector: 'input[mdDatepicker], input[matDatepicker]',
                providers: [MD_DATEPICKER_VALUE_ACCESSOR, MD_DATEPICKER_VALIDATORS],
                host: {
                    '[attr.aria-haspopup]': 'true',
                    '[attr.aria-owns]': '(_datepicker?.opened && _datepicker.id) || null',
                    '[attr.min]': 'min ? _dateAdapter.getISODateString(min) : null',
                    '[attr.max]': 'max ? _dateAdapter.getISODateString(max) : null',
                    '[disabled]': 'disabled',
                    '(input)': '_onInput($event.target.value)',
                    '(change)': '_onChange()',
                    '(blur)': '_onTouched()',
                    '(keydown)': '_onKeydown($event)',
                },
                exportAs: 'mdDatepickerInput',
            },] }
];
/** @nocollapse */
MdDatepickerInput.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: DateAdapter, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_FORMATS,] }] },
    { type: MatFormField, decorators: [{ type: Optional }] }
];
MdDatepickerInput.propDecorators = {
    mdDatepicker: [{ type: Input }],
    matDatepicker: [{ type: Input }],
    mdDatepickerFilter: [{ type: Input }],
    matDatepickerFilter: [{ type: Input }],
    value: [{ type: Input }],
    min: [{ type: Input }],
    max: [{ type: Input }],
    disabled: [{ type: Input }],
    dateChange: [{ type: Output }],
    dateInput: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    MdDatepickerInput.prototype._datepicker;
    /** @type {?} */
    MdDatepickerInput.prototype._dateFilter;
    /**
     * The value of the input.
     * @type {?}
     */
    MdDatepickerInput.prototype._value;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._min;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._max;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._disabled;
    /**
     * Emits when a `change` event is fired on this `<input>`.
     * @type {?}
     */
    MdDatepickerInput.prototype.dateChange;
    /**
     * Emits when an `input` event is fired on this `<input>`.
     * @type {?}
     */
    MdDatepickerInput.prototype.dateInput;
    /**
     * Emits when the value changes (either due to user input or programmatic change).
     * @type {?}
     */
    MdDatepickerInput.prototype._valueChange;
    /** @type {?} */
    MdDatepickerInput.prototype._onTouched;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._cvaOnChange;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._validatorOnChange;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._datepickerSubscription;
    /**
     * The form control validator for whether the input parses.
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._parseValidator;
    /**
     * The form control validator for the min date.
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._minValidator;
    /**
     * The form control validator for the max date.
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._maxValidator;
    /**
     * The form control validator for the date filter.
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._filterValidator;
    /**
     * The combined form control validator for this input.
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._validator;
    /**
     * Whether the last value set on the input was valid.
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._lastValueValid;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._dateFormats;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerInput.prototype._mdFormField;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/datepicker-toggle.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @template D
 */
class MdDatepickerToggle {
    /**
     * @param {?} _intl
     * @param {?} changeDetectorRef
     */
    constructor(_intl, changeDetectorRef) {
        this._intl = _intl;
        this._intlChanges = _intl.changes.subscribe((/**
         * @return {?}
         */
        () => changeDetectorRef.markForCheck()));
    }
    /**
     * Whether the toggle button is disabled.
     * @return {?}
     */
    get disabled() {
        return this._disabled === undefined ? this.datepicker.disabled : this._disabled;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this._disabled = coerceBooleanProperty(value);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._intlChanges.unsubscribe();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _open(event) {
        if (this.datepicker && !this.disabled) {
            this.datepicker.open();
            event.stopPropagation();
        }
    }
}
MdDatepickerToggle.decorators = [
    { type: Component, args: [{
                selector: 'md-datepicker-toggle, mat-datepicker-toggle',
                template: "<button mat-icon-button type=\"button\" [attr.aria-label]=\"_intl.openCalendarLabel\"\r\n        [disabled]=\"disabled\" (click)=\"_open($event)\">\r\n  <!-- <mat-icon>\r\n    <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" width=\"100%\" height=\"100%\"\r\n         fill=\"currentColor\" style=\"vertical-align: top\" focusable=\"false\">\r\n      <path d=\"M0 0h24v24H0z\" fill=\"none\"/>\r\n      <path d=\"M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zM7 10h5v5H7z\"/>\r\n    </svg>\r\n  </mat-icon> -->\r\n</button>\r\n",
                host: {
                    'class': 'mat-datepicker-toggle',
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
MdDatepickerToggle.ctorParameters = () => [
    { type: MdDatepickerIntl },
    { type: ChangeDetectorRef }
];
MdDatepickerToggle.propDecorators = {
    datepicker: [{ type: Input, args: ['for',] }],
    disabled: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    MdDatepickerToggle.prototype._intlChanges;
    /**
     * Datepicker instance that the button will toggle.
     * @type {?}
     */
    MdDatepickerToggle.prototype.datepicker;
    /**
     * @type {?}
     * @private
     */
    MdDatepickerToggle.prototype._disabled;
    /** @type {?} */
    MdDatepickerToggle.prototype._intl;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/locale_changer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * An internal component used to display a single year in the datepicker.
 * \@docs-private
 * @template D
 */
class MdLocaleChanger {
    /**
     * @param {?} _dateAdapter
     * @param {?} _dateFormats
     */
    constructor(_dateAdapter, _dateFormats) {
        this._dateAdapter = _dateAdapter;
        this._dateFormats = _dateFormats;
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        if (!this._dateFormats) {
            throw createMissingDateImplError('MD_DATE_FORMATS');
        }
        this.any = (/** @type {?} */ (_dateAdapter));
        this.mustDestroy = ((/** @type {?} */ (_dateFormats)))._lastChanges.subscribe((/**
         * @param {?} v
         * @return {?}
         */
        (v) => {
            this.lang = v;
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.mustDestroy.unsubscribe();
    }
    // change locale function()
    /**
     * @return {?}
     */
    _changeLocale() {
        this.any.changeLocale();
    }
}
MdLocaleChanger.decorators = [
    { type: Component, args: [{
                moduleId: module.id,
                selector: 'md-locale-changer',
                template: `
        <mat-button-toggle-group #group="matButtonToggleGroup" (change)="_changeLocale()" [value]="lang">
            <mat-button-toggle value="fa">
                
            </mat-button-toggle>
            <mat-button-toggle value="en">
                
            </mat-button-toggle>
        </mat-button-toggle-group>
    `,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".mat-button-toggle-group .mat-button-toggle{font-size:11px}.mat-button-toggle-group .mat-button-toggle .mat-button-toggle-label-content{line-height:26px}"]
            }] }
];
/** @nocollapse */
MdLocaleChanger.ctorParameters = () => [
    { type: DateAdapter, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_FORMATS,] }] }
];
if (false) {
    /** @type {?} */
    MdLocaleChanger.prototype.any;
    /** @type {?} */
    MdLocaleChanger.prototype.lang;
    /** @type {?} */
    MdLocaleChanger.prototype.mustDestroy;
    /** @type {?} */
    MdLocaleChanger.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdLocaleChanger.prototype._dateFormats;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/injectors.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MAT_DATE_LOCALE = new InjectionToken('MAT_DATE_LOCALE');

/**
 * @fileoverview added by tsickle
 * Generated from: lib/jalali-date-adapter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const minmin = jMoment;
class JalaliMomentDateAdapter extends DateAdapter {
    /**
     * @param {?} matDateLocale
     */
    constructor(matDateLocale) {
        super();
        this.setLocale(matDateLocale || minmin.locale('fa'));
        minmin.loadPersian();
    }
    /**
     * @return {?}
     */
    invalid() {
        return (/** @type {?} */ ('hi'));
    }
    /**
     * @param {?} date
     * @return {?}
     */
    toIso8601(date) {
        return 'hi';
    }
    /**
     * returns year in jalali calendar system.
     * @param {?} date
     * @return {?}
     */
    getYear(date) {
        return this.clone(date).jYear();
    }
    /**
     * returns month in jalali calendar system.
     * @param {?} date
     * @return {?}
     */
    getMonth(date) {
        return this.clone(date).jMonth();
    }
    /**
     * returns day in jalali calendar system.
     * @param {?} date
     * @return {?}
     */
    getDate(date) {
        return this.clone(date).jDate();
    }
    /**
     * returns Day Of Week in jalali calendar system.
     * @param {?} date
     * @return {?}
     */
    getDayOfWeek(date) {
        return this.clone(date).day();
    }
    /**
     * returns Month Names in jalali calendar system.
     * most of the time we use long format. short or narrow format for month names is a little odd.
     * @param {?} style
     * @return {?}
     */
    getMonthNames(style) {
        /** @type {?} */
        const r = [];
        minmin.loadPersian({ dialect: 'persian', usePersianDigits: true });
        /** @type {?} */
        const l = minmin().localeData();
        if (style === 'long' || style === 'short') {
            Object.keys(l).forEach((/**
             * @param {?} key
             * @return {?}
             */
            key => {
                if (key === '_jMonths') {
                    /** @type {?} */
                    const values = l[key];
                    for (let index = 0; index < values.length; index++) {
                        r.push(values[index]);
                    }
                }
            }));
        }
        else {
            Object.keys(l).forEach((/**
             * @param {?} key
             * @return {?}
             */
            key => {
                if (key === '_jMonthsShort') {
                    /** @type {?} */
                    const values = l[key];
                    for (let index = 0; index < values.length; index++) {
                        r.push(values[index]);
                    }
                }
            }));
        }
        return r;
        // switch (style) {
        //     case 'long':
        //         const mlong = (minmin().localeData() as any)._jMonths.slice(0);
        //         console.log(mlong);
        //         return mlong;
        //     case 'short':
        //         const mshort = (minmin().localeData() as any)._jMonthsShort.slice(0);
        //         console.log('mshort: ', mshort);
        //         return mshort;
        //     case 'narrow':
        //         return (minmin().localeData() as any)._jMonthsShort.slice(0);
        // }
    }
    /**
     * borrowed from angular material code.
     * @return {?}
     */
    getDateNames() {
        return this.range(31, (/**
         * @param {?} i
         * @return {?}
         */
        i => String(i + 1)));
    }
    /**
     * returns Day Of Week names in jalali calendar system.
     * @param {?} style
     * @return {?}
     */
    getDayOfWeekNames(style) {
        switch (style) {
            case 'long':
                return minmin().localeData().weekdays().slice(0);
            case 'short':
                return minmin().localeData().weekdaysShort().slice(0);
            case 'narrow':
                return minmin().localeData().weekdaysMin().slice(0);
        }
    }
    /**
     * returns year in jalali calendar system.
     * @param {?} date
     * @return {?}
     */
    getYearName(date) {
        return this.clone(date).jYear().toString();
    }
    /**
     * returns first day of week in jalali calendar system.
     * first day of week is saturday, 
     * @return {?}
     */
    getFirstDayOfWeek() {
        return 6;
    }
    /**
     * returns Number of Days In Month in jalali calendar system.
     * @param {?} date
     * @param {?=} fa
     * @return {?}
     */
    getNumDaysInMonth(date, fa) {
        if (date['_d']) {
            return minmin.jDaysInMonth(this.getYear(date['_d']), this.getMonth(date['_d']));
        }
        return minmin.jDaysInMonth(this.getYear(date), this.getMonth(date));
    }
    /**
     * @param {?} date
     * @return {?}
     */
    clone(date) {
        // return date.clone().locale(this.locale);
        return minmin(date);
    }
    /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    createDate(year, month, date) {
        return minmin().jYear(year).jMonth(month).jDate(date);
    }
    /**
     * @return {?}
     */
    today() {
        return minmin();
    }
    /**
     * @param {?} value
     * @param {?} parseFormat
     * @return {?}
     */
    parse(value, parseFormat) {
        if (value && typeof value === 'string') {
            return minmin(value, parseFormat, this.locale);
        }
        return value ? minmin(value).locale(this.locale) : null;
    }
    /**
     * @param {?} date
     * @param {?} displayFormat
     * @return {?}
     */
    format(date, displayFormat) {
        return this.clone(date).format(displayFormat);
    }
    /**
     * @param {?} date
     * @param {?} years
     * @return {?}
     */
    addCalendarYears(date, years) {
        return this.clone(date).add(years, 'jYear');
    }
    /**
     * @param {?} date
     * @param {?} months
     * @return {?}
     */
    addCalendarMonths(date, months) {
        return this.clone(date).add((/** @type {?} */ (months)), 'jmonth');
    }
    /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    addCalendarDays(date, days) {
        return this.clone(date).add(days, 'day');
    }
    /**
     * @param {?} date
     * @return {?}
     */
    getISODateString(date) {
        return this.clone(date).format('jYYYY-jMM-jDD');
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    isDateInstance(obj) {
        return minmin.isMoment(obj);
    }
    /**
     * @param {?} date
     * @return {?}
     */
    isValid(date) {
        return this.clone(date).isValid();
    }
    /**
     * @template T
     * @param {?} length
     * @param {?} valueFunction
     * @return {?}
     */
    range(length, valueFunction) {
        /** @type {?} */
        const valuesArray = Array(length);
        for (let i = 0; i < length; i++) {
            valuesArray[i] = valueFunction(i);
        }
        return valuesArray;
    }
}
/** @nocollapse */
JalaliMomentDateAdapter.ctorParameters = () => [
    { type: String, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_LOCALE,] }] }
];

/**
 * @fileoverview added by tsickle
 * Generated from: lib/object-extend.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Extends an object with the *enumerable* and *own* properties of one or more source objects,
 * similar to Object.assign.
 *
 * @param {?} dest The object which will have properties copied to it.
 * @param {...?} sources The source objects from which properties will be copied.
 * @return {?}
 */
function extendObject(dest, ...sources) {
    if (dest == null) {
        throw TypeError('Cannot convert undefined or null to object');
    }
    for (let source of sources) {
        if (source != null) {
            for (let key in source) {
                if (source.hasOwnProperty(key)) {
                    dest[key] = source[key];
                }
            }
        }
    }
    return dest;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/moment-date-adapter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TODO(mmalerba): Remove when we no longer support safari 9.
/**
 * Whether the browser supports the Intl API.
 * @type {?}
 */
const SUPPORTS_INTL_API = typeof Intl != 'undefined';
/**
 * The default month names to use if Intl API is not available.
 * @type {?}
 */
const DEFAULT_MONTH_NAMES = {
    'long': [
        'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September',
        'October', 'November', 'December'
    ],
    'short': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
    'narrow': ['J', 'F', 'M', 'A', 'M', 'J', 'J', 'A', 'S', 'O', 'N', 'D']
};
const 0 = /**
 * @param {?} i
 * @return {?}
 */
i => String(i + 1);
/**
 * The default date names to use if Intl API is not available.
 * @type {?}
 */
const DEFAULT_DATE_NAMES = range(31, (0));
/**
 * The default day of the week names to use if Intl API is not available.
 * @type {?}
 */
const DEFAULT_DAY_OF_WEEK_NAMES = {
    'long': ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
    'short': ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
    'narrow': ['S', 'M', 'T', 'W', 'T', 'F', 'S']
};
/**
 * Creates an array and fills it with values.
 * @template T
 * @param {?} length
 * @param {?} valueFunction
 * @return {?}
 */
function range(length, valueFunction) {
    /** @type {?} */
    const valuesArray = Array(length);
    for (let i = 0; i < length; i++) {
        valuesArray[i] = valueFunction(i);
    }
    return valuesArray;
}
/**
 * Adapts the native JS Date for use with cdk-based components that work with dates.
 */
class NativeDateAdapter extends DateAdapter {
    /**
     * @param {?} localeId
     */
    constructor(localeId) {
        super();
        this.lang = 'fa';
        this.langSubject = new Subject();
        this.JDateAdapter = new JalaliMomentDateAdapter('fa');
        /**
         * Whether to use `timeZone: 'utc'` with `Intl.DateTimeFormat` when formatting dates.
         * Without this `Intl.DateTimeFormat` sometimes chooses the wrong timeZone, which can throw off
         * the result. (e.g. in the en-US locale `new Date(1800, 7, 14).toLocaleDateString()`
         * will produce `'8/13/1800'`.
         */
        this.useUtcForDisplay = true;
        super.setLocale(localeId);
    }
    /**
     * @return {?}
     */
    invalid() {
        return new Date();
    }
    /**
     * @param {?} date
     * @return {?}
     */
    toIso8601(date) {
        return 'hi';
    }
    /**
     * @param {?} a
     * @param {?} b
     * @return {?}
     */
    sameDate(a, b) {
        return false;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    setLocale(v) {
        super.setLocale(v);
        if (this.lang != 'fa') {
            this.lang = 'fa';
            this.langSubject.next('fa');
        }
        else {
            this.lang = 'en';
            this.langSubject.next('en');
        }
    }
    /**
     * @return {?}
     */
    getLang() {
        return this.langSubject;
    }
    /**
     * @return {?}
     */
    changeLocale() {
        if (this.lang != 'fa') {
            this.setLocale('fa');
        }
        else {
            this.setLocale('en');
        }
    }
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    getYear(date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getYear((/** @type {?} */ (date)));
        }
        if (date._d) {
            return date._d.getFullYear();
        }
        if (!date.getFullYear) {
            debugger;
        }
        return date.getFullYear();
    }
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    getMonth(date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getMonth((/** @type {?} */ (date)));
        }
        if (date._d) {
            return date._d.getMonth();
        }
        return date.getMonth();
    }
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    getDate(date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDate((/** @type {?} */ (date)));
        }
        if (date._d) {
            return date._d.getDate();
        }
        return date.getDate();
    }
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    getDayOfWeek(date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDayOfWeek((/** @type {?} */ (date)));
        }
        return date.getDay();
    }
    /**
     * @param {?} style
     * @param {?=} force
     * @return {?}
     */
    getMonthNames(style, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getMonthNames(style);
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            const dtf = new Intl.DateTimeFormat(this.locale, { month: style });
            return range(12, (/**
             * @param {?} i
             * @return {?}
             */
            i => this._stripDirectionalityCharacters(dtf.format(new Date(2017, i, 1)))));
        }
        return DEFAULT_MONTH_NAMES[style];
    }
    /**
     * @param {?=} force
     * @return {?}
     */
    getDateNames(force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDateNames();
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            const dtf = new Intl.DateTimeFormat(this.locale, { day: 'numeric' });
            return range(31, (/**
             * @param {?} i
             * @return {?}
             */
            i => this._stripDirectionalityCharacters(dtf.format(new Date(2017, 0, i + 1)))));
        }
        return DEFAULT_DATE_NAMES;
    }
    /**
     * @param {?} style
     * @param {?=} force
     * @return {?}
     */
    getDayOfWeekNames(style, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getDayOfWeekNames(style);
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            const dtf = new Intl.DateTimeFormat(this.locale, { weekday: style });
            return range(7, (/**
             * @param {?} i
             * @return {?}
             */
            i => this._stripDirectionalityCharacters(dtf.format(new Date(2017, 0, i + 1)))));
        }
        return DEFAULT_DAY_OF_WEEK_NAMES[style];
    }
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    getYearName(date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getYearName((/** @type {?} */ (date)));
        }
        if (SUPPORTS_INTL_API) {
            /** @type {?} */
            const dtf = new Intl.DateTimeFormat(this.locale, { year: 'numeric' });
            return this._stripDirectionalityCharacters(dtf.format(date));
        }
        return String(this.getYear(date));
    }
    /**
     * @param {?=} force
     * @return {?}
     */
    getFirstDayOfWeek(force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getFirstDayOfWeek();
        }
        // We can't tell using native JS Date what the first day of the week is, we default to Sunday.
        return 0;
    }
    /**
     * @param {?} date
     * @param {?=} force
     * @return {?}
     */
    getNumDaysInMonth(date, force) {
        if (this.lang == 'fa' && !force) {
            return this.JDateAdapter.getNumDaysInMonth((/** @type {?} */ (date)), 'fa');
        }
        return this.getDate(this._createDateWithOverflow(this.getYear(date), this.getMonth(date) + 1, 0));
    }
    /**
     * @param {?} date
     * @return {?}
     */
    clone(date) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ ((this.JDateAdapter.clone((/** @type {?} */ (date))))));
        }
        return this.createDate(this.getYear(date), this.getMonth(date), this.getDate(date));
    }
    /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    createDate(year, month, date) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ ((this.JDateAdapter.createDate(year, month, date))));
        }
        // Check for invalid month and date (except upper bound on date which we have to check after
        // creating the Date).
        if (month < 0 || month > 11) {
            throw Error(`Invalid month index "${month}". Month index has to be between 0 and 11.`);
        }
        if (date < 1) {
            throw Error(`Invalid date "${date}". Date has to be greater than 0.`);
        }
        /** @type {?} */
        const result = this._createDateWithOverflow(year, month, date);
        // Check that the date wasn't above the upper bound for the month, causing the month to overflow
        if (result.getMonth() != month) {
            throw Error(`Invalid date "${date}" for month with index "${month}".`);
        }
        return result;
    }
    /**
     * @return {?}
     */
    today() {
        if (this.lang == 'fa') {
            return (/** @type {?} */ ((this.JDateAdapter.today())));
        }
        return new Date();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    parse(value) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.parse(value, [])));
        }
        // We have no way using the native JS Date to set the parse format or locale, so we ignore these
        // parameters.
        if (typeof value == 'number') {
            return new Date(value);
        }
        return value ? new Date(Date.parse(value)) : null;
    }
    /**
     * @param {?} date
     * @param {?} displayFormat
     * @return {?}
     */
    format(date, displayFormat) {
        if (this.lang == 'fa') {
            return this.JDateAdapter.format((/** @type {?} */ (date)), displayFormat);
        }
        if (!this.isValid(date)) {
            throw Error('NativeDateAdapter: Cannot format invalid date.');
        }
        if (SUPPORTS_INTL_API) {
            if (this.useUtcForDisplay) {
                if (date._d) {
                    date = new Date(Date.UTC(date._d.getFullYear(), date._d.getMonth(), date._d.getDate(), date._d.getHours(), date._d.getMinutes(), date._d.getSeconds(), date._d.getMilliseconds()));
                    displayFormat = extendObject({}, displayFormat, { timeZone: 'utc' });
                }
                else {
                    date = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds()));
                    displayFormat = extendObject({}, displayFormat, { timeZone: 'utc' });
                }
            }
            /** @type {?} */
            const dtf = new Intl.DateTimeFormat(this.locale, displayFormat);
            return this._stripDirectionalityCharacters(dtf.format(date));
        }
        return this._stripDirectionalityCharacters(date.toDateString());
    }
    /**
     * @param {?} date
     * @param {?} years
     * @return {?}
     */
    addCalendarYears(date, years) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.addCalendarYears((/** @type {?} */ (date)), years)));
        }
        return this.addCalendarMonths(date, years * 12);
    }
    /**
     * @param {?} date
     * @param {?} months
     * @return {?}
     */
    addCalendarMonths(date, months) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.addCalendarMonths((/** @type {?} */ (date)), months)));
        }
        /** @type {?} */
        let newDate = this._createDateWithOverflow(this.getYear(date), this.getMonth(date) + months, this.getDate(date));
        // It's possible to wind up in the wrong month if the original month has more days than the new
        // month. In this case we want to go to the last day of the desired month.
        // Note: the additional + 12 % 12 ensures we end up with a positive number, since JS % doesn't
        // guarantee this.
        if (this.getMonth(newDate) != ((this.getMonth(date) + months) % 12 + 12) % 12) {
            newDate = this._createDateWithOverflow(this.getYear(newDate), this.getMonth(newDate), 0);
        }
        return newDate;
    }
    /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    addCalendarDays(date, days) {
        if (this.lang == 'fa') {
            return (/** @type {?} */ (this.JDateAdapter.addCalendarDays((/** @type {?} */ (date)), days)));
        }
        return this._createDateWithOverflow(this.getYear(date), this.getMonth(date), this.getDate(date) + days);
    }
    /**
     * @param {?} date
     * @return {?}
     */
    getISODateString(date) {
        if (this.lang == 'fa') {
            return this.JDateAdapter.getISODateString((/** @type {?} */ (date)));
        }
        return [
            date.getUTCFullYear(),
            this._2digit(date.getUTCMonth() + 1),
            this._2digit(date.getUTCDate())
        ].join('-');
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    isDateInstance(obj) {
        return obj instanceof Date;
    }
    /**
     * @param {?} date
     * @return {?}
     */
    isValid(date) {
        if (date._d) {
            return !isNaN(date._d.getTime());
        }
        return !isNaN(date.getTime());
    }
    /**
     * Creates a date but allows the month and date to overflow.
     * @private
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    _createDateWithOverflow(year, month, date) {
        /** @type {?} */
        const result = new Date(year, month, date);
        // We need to correct for the fact that JS native Date treats years in range [0, 99] as
        // abbreviations for 19xx.
        if (year >= 0 && year < 100) {
            result.setFullYear(this.getYear(result) - 1900);
        }
        return result;
    }
    /**
     * Pads a number to make it two digits.
     * @private
     * @param {?} n The number to pad.
     * @return {?} The padded number.
     */
    _2digit(n) {
        return ('00' + n).slice(-2);
    }
    /**
     * Strip out unicode LTR and RTL characters. Edge and IE insert these into formatted dates while
     * other browsers do not. We remove them to make output consistent and because they interfere with
     * date parsing.
     * @private
     * @param {?} str The string to strip direction characters from.
     * @return {?} The stripped string.
     */
    _stripDirectionalityCharacters(str) {
        return str.replace(/[\u200e\u200f]/g, '');
    }
}
NativeDateAdapter.decorators = [
    { type: Injectable }
];
/** @nocollapse */
NativeDateAdapter.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [LOCALE_ID,] }] }
];
if (false) {
    /** @type {?} */
    NativeDateAdapter.prototype.lang;
    /** @type {?} */
    NativeDateAdapter.prototype.langSubject;
    /** @type {?} */
    NativeDateAdapter.prototype.JDateAdapter;
    /**
     * Whether to use `timeZone: 'utc'` with `Intl.DateTimeFormat` when formatting dates.
     * Without this `Intl.DateTimeFormat` sometimes chooses the wrong timeZone, which can throw off
     * the result. (e.g. in the en-US locale `new Date(1800, 7, 14).toLocaleDateString()`
     * will produce `'8/13/1800'`.
     * @type {?}
     */
    NativeDateAdapter.prototype.useUtcForDisplay;
}
class dateFormat {
    /**
     * @param {?} _dateAdapter
     */
    constructor(_dateAdapter) {
        this.parse = {
            dateInput: 'l',
        };
        this.display = {
            dateInput: 'jYYYY/jMM/jDD',
            monthYearLabel: 'jYYYY jMMMM',
            dateA11yLabel: 'jYYYY/jMM/jDD',
            monthYearA11yLabel: 'jYYYY jMMMM'
        };
        this._getChanges = new Subject();
        this._lastChanges = new BehaviorSubject('fa');
        /** @type {?} */
        const dateAdapter = (/** @type {?} */ (_dateAdapter));
        dateAdapter.getLang().subscribe((/**
         * @param {?} v
         * @return {?}
         */
        v => {
            if (v == 'fa') {
                this.parse = {
                    dateInput: 'l',
                };
                this.display = {
                    dateInput: 'jYYYY/jMM/jDD',
                    monthYearLabel: 'jYYYY jMMMM',
                    dateA11yLabel: 'jYYYY/jMM/jDD',
                    monthYearA11yLabel: 'jYYYY jMMMM'
                };
            }
            else {
                this.parse = {
                    dateInput: null,
                };
                this.display = {
                    dateInput: { year: 'numeric', month: 'numeric', day: 'numeric' },
                    monthYearLabel: { year: 'numeric', month: 'short' },
                    dateA11yLabel: { year: 'numeric', month: 'long', day: 'numeric' },
                    monthYearA11yLabel: { year: 'numeric', month: 'long' },
                };
            }
            this._lastChanges.next(v);
            this._getChanges.next(v);
        }), (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
        }), (/**
         * @return {?}
         */
        () => {
        }));
    }
    /**
     * @return {?}
     */
    getChanges() {
        return this._getChanges;
    }
}
if (false) {
    /** @type {?} */
    dateFormat.prototype.parse;
    /** @type {?} */
    dateFormat.prototype.display;
    /** @type {?} */
    dateFormat.prototype._getChanges;
    /** @type {?} */
    dateFormat.prototype._lastChanges;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MatDatepickerModulePersian {
}
MatDatepickerModulePersian.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    MatButtonModule,
                    MatButtonToggleModule,
                    MatDialogModule,
                    MatIconModule,
                    OverlayModule,
                    MatOptionModule,
                    MatSelectModule,
                    // StyleModule,
                    A11yModule,
                ],
                exports: [
                    MdCalendar,
                    MdCalendarBody,
                    MdDatepicker,
                    MatButtonToggleModule,
                    MdDatepickerInput,
                    MdDatepickerToggle,
                    MdMonthView,
                    MdYearView,
                ],
                declarations: [
                    MdCalendar,
                    MdCalendarBody,
                    MdLocaleChanger,
                    MdDatepicker,
                    MdDatepickerContent,
                    MdDatepickerInput,
                    MdDatepickerToggle,
                    MdMonthView,
                    MdYearView,
                ],
                providers: [
                    MdDatepickerIntl,
                    MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER,
                    { provide: DateAdapter, useClass: NativeDateAdapter }, {
                        provide: MAT_DATE_FORMATS,
                        useClass: dateFormat,
                        deps: [DateAdapter]
                    }
                ],
                entryComponents: [
                    MdDatepickerContent,
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * Generated from: lib/public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ngx-persian-datepicker.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MD_DATEPICKER_SCROLL_STRATEGY, MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER, MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER_FACTORY, MD_DATEPICKER_VALIDATORS, MD_DATEPICKER_VALUE_ACCESSOR, MatDatepickerModulePersian, MdCalendar, MdCalendarBody, MdCalendarCell, MdDatepicker, MdDatepickerContent, MdDatepickerInput, MdDatepickerInputEvent, MdDatepickerIntl, MdDatepickerToggle, MdMonthView, MdYearView, MdLocaleChanger as a, NativeDateAdapter as b, dateFormat as c };
