// ngx-persian-datepicker/esm2015/lib/locale_changer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/locale_changer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { ChangeDetectionStrategy, Component, Inject, Optional, ViewEncapsulation } from '@angular/core';
import { createMissingDateImplError } from './datepicker-errors';
import { DateAdapter, MAT_DATE_FORMATS } from '@angular/material';
/**
 * An internal component used to display a single year in the datepicker.
 * \@docs-private
 * @template D
 */
export class MdLocaleChanger {
    /**
     * @param {?} _dateAdapter
     * @param {?} _dateFormats
     */
    constructor(_dateAdapter, _dateFormats) {
        this._dateAdapter = _dateAdapter;
        this._dateFormats = _dateFormats;
        if (!this._dateAdapter) {
            throw createMissingDateImplError('DateAdapter');
        }
        if (!this._dateFormats) {
            throw createMissingDateImplError('MD_DATE_FORMATS');
        }
        this.any = (/** @type {?} */ (_dateAdapter));
        this.mustDestroy = ((/** @type {?} */ (_dateFormats)))._lastChanges.subscribe((/**
         * @param {?} v
         * @return {?}
         */
        (v) => {
            this.lang = v;
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.mustDestroy.unsubscribe();
    }
    // change locale function()
    /**
     * @return {?}
     */
    _changeLocale() {
        this.any.changeLocale();
    }
}
MdLocaleChanger.decorators = [
    { type: Component, args: [{
                moduleId: module.id,
                selector: 'md-locale-changer',
                template: `
        <mat-button-toggle-group #group="matButtonToggleGroup" (change)="_changeLocale()" [value]="lang">
            <mat-button-toggle value="fa">
                شمسی
            </mat-button-toggle>
            <mat-button-toggle value="en">
                میلادی
            </mat-button-toggle>
        </mat-button-toggle-group>
    `,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".mat-button-toggle-group .mat-button-toggle{font-size:11px}.mat-button-toggle-group .mat-button-toggle .mat-button-toggle-label-content{line-height:26px}"]
            }] }
];
/** @nocollapse */
MdLocaleChanger.ctorParameters = () => [
    { type: DateAdapter, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_FORMATS,] }] }
];
if (false) {
    /** @type {?} */
    MdLocaleChanger.prototype.any;
    /** @type {?} */
    MdLocaleChanger.prototype.lang;
    /** @type {?} */
    MdLocaleChanger.prototype.mustDestroy;
    /** @type {?} */
    MdLocaleChanger.prototype._dateAdapter;
    /**
     * @type {?}
     * @private
     */
    MdLocaleChanger.prototype._dateFormats;
}
